<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>ADMIXTOOLS 2 Torial • admixtools</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><script src="../extra.js"></script><meta property="og:title" content="ADMIXTOOLS 2 Torial">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-155330476-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-155330476-1');
</script>
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">admixtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/admixtools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/fstats.html">f-statistics</a>
    </li>
    <li>
      <a href="../articles/graphs.html">Admixture graphs</a>
    </li>
    <li>
      <a href="../articles/io.html">Genotype data</a>
    </li>
    <li>
      <a href="../articles/parallel.html">Parallelization</a>
    </li>
    <li>
      <a href="../articles/plotting.html">Plotting</a>
    </li>
    <li>
      <a href="../articles/resampling.html">Standard errors</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="admixtools_files/htmlwidgets-1.3/htmlwidgets.js"></script><script src="admixtools_files/plotly-binding-4.9.1/plotly.js"></script><script src="admixtools_files/typedarray-0.1/typedarray.min.js"></script><script src="admixtools_files/jquery-1.11.3/jquery.min.js"></script><link href="admixtools_files/crosstalk-1.0.0/css/crosstalk.css" rel="stylesheet">
<script src="admixtools_files/crosstalk-1.0.0/js/crosstalk.min.js"></script><link href="admixtools_files/plotly-htmlwidgets-css-1.49.4/plotly-htmlwidgets.css" rel="stylesheet">
<script src="admixtools_files/plotly-main-1.49.4/plotly-latest.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>ADMIXTOOLS 2 Torial</h1>
                        <h4 class="author">Robert Maier</h4>
            
            <h4 class="date">2020-08-12</h4>
      
      
      <div class="hidden name"><code>admixtools.Rmd</code></div>

    </div>

    
    
<p>This tutorial gives an overview of the basic workflow for computing <em>f</em>-statistics, and using <em>qpWave</em>, <em>qpAdm</em>, and <em>qpGraph</em>. Documentation for each ADMIXTOOLS function can be found under <em>Reference</em>, and more detailed information about specific topics under <em>Articles</em>.</p>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>Admixtools is a set of programs used to infer population histories from f-statistics computed from genetic data. ADMIXTOOLS 2 divides the computation into two steps:</p>
<ol style="list-style-type: decimal">
<li>Computing <span class="math inline">\(f_2\)</span>-statistics and storing them on disk. This can be slow since it accesses the genotype data.</li>
<li>Using <span class="math inline">\(f_2\)</span>-statistics to fit models. This is fast because <span class="math inline">\(f_2\)</span>-statistics are very compact compared to genotype data.</li>
</ol>
<p>ADMIXTOOLS 2 can either be run through the <code>R</code> command line, or through a browser application that can be launched like this:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(admixtools)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="../reference/run_shiny_admixtools.html">run_shiny_admixtools</a></span>()</a></code></pre></div>
<p>The documentation of the browser application can be accessed through the <code>?</code> in the upper right corner of the application window. This website documents the <code>R</code> command line interface.</p>
<p>For the examples below the following R packages need to be loaded.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(admixtools)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(magrittr)</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(tidyverse)</a></code></pre></div>
<p><br></p>
</div>
<div id="f-statistics" class="section level2">
<h2 class="hasAnchor">
<a href="#f-statistics" class="anchor"></a><em>f</em>-statistics</h2>
<p><span class="math inline">\(f_2\)</span>, <span class="math inline">\(f_3\)</span>, and <span class="math inline">\(f_4\)</span> describe how populations are related to one another, and they form the basis of ADMIXTOOLS. Here we briefly describe what they are and how to compute them. Two excellent papers on f-statistics can be found <a href="https://www.genetics.org/content/192/3/1065">here</a> and <a href="https://www.genetics.org/content/202/4/1485">here</a>.</p>
<p><span class="math inline">\(f_2\)</span> is the expected squared difference in allele frequencies between two populations <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span>. (Allele frequencies are denoted in lower case letters here.) We can estimate it by averaging over <span class="math inline">\(M\)</span> SNPs: <span class="math display">\[f_2(A,B) = \frac{1}{M} \sum_{j=1}^M(a_{j} - b_{j})^2\]</span></p>
<p><span class="math inline">\(f_4\)</span> is the covariance of allele frequency differences between two pairs of populations, and at the same time the sum of four f2-statistics: <span class="math display">\[
\begin{equation}
\begin{aligned}
f_4(A, B; C, D) &amp;= \frac{1}{M}\sum_{j=1}^M(a_{j} - b_{j})(c_{j} - d_{j}) \\ &amp;= \frac{1}{2}(f_2(A, D) + f_2(B, C) - f_2(A, C) - f_2(B, D) )
\label{eq:f42}
\end{aligned}
\end{equation}
\]</span></p>
<p><span class="math inline">\(f_3\)</span> is the covariance of allele frequency differences between two pairs of populations where one population is the same on both sides. <span class="math inline">\(f_3\)</span> is thus a special case of <span class="math inline">\(f_4\)</span>, and can also be written as a sum of <span class="math inline">\(f_2\)</span>.</p>
<p><span class="math display">\[
\begin{aligned}
f_3(A; B, C) &amp;= f_4(A, B; A, C) \\ &amp;= \frac{1}{2} (f_2(A, B) + f_2(A, C) - f_2(B, C))
\end{aligned}
\]</span></p>
<p>In practice, the estimation of f-statistics can be more complicated, as it needs to account for low sample counts, missing data, and differences in ploidy, as described <a href="fstats.html">here</a>.</p>
<p>In ADMIXTOOLS 2, f2-statistics are the foundation for all further analyses. They can be computed from genotype data and saved to disk with this command:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">prefix =<span class="st"> '/path/to/geno'</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">my_f2_dir =<span class="st"> '/store/f2data/here/'</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3"></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="kw"><a href="../reference/extract_f2.html">extract_f2</a></span>(prefix, my_f2_dir)</a></code></pre></div>
<p>This will look for genotype files in <code>packedancestrymap</code> or <code>PLINK</code> format, compute allele frequencies and blocked f2 statistics for all pairs of populations defined in the <code>.ind</code> or <code>.fam</code> file, and write them to <code>my_f2_dir</code>. It is also possible to extract only a subset of the samples or populations by passing IDs to the <code>inds</code> and <code>pops</code> arguments in <code>extract_f2</code>. To get a description of the arguemnts and to see examples of how to use it, type</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">?extract_f2</a></code></pre></div>
<p>Once finished, f2-statistics for the populations of interest can be loaded like this.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">f2_blocks =<span class="st"> </span><span class="kw"><a href="../reference/f2_from_precomp.html">f2_from_precomp</a></span>(my_f2_dir)</a></code></pre></div>
<p>Or you can load only a subset of the populations:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">mypops =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'Denisova.DG'</span>, <span class="st">'Altai_Neanderthal.DG'</span>, <span class="st">'Vindija.DG'</span>)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">f2_blocks =<span class="st"> </span><span class="kw"><a href="../reference/f2_from_precomp.html">f2_from_precomp</a></span>(my_f2_dir, <span class="dt">pops =</span> mypops)</a></code></pre></div>
<p><code>f2_blocks</code> is now a 3d-array with f2-statistics for each population pair (along dimensions 1 and 2) and each SNP block (along the 3rd dimension).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(f2_blocks)</a></code></pre></div>
<pre><code>## [1]   7   7 708</code></pre>
<p>Separate estimates for each SNP block make it possible to compute jackknife or bootstrap standard errors for any f-statistic, and for any other estimated parameter.</p>
<p><code>f2_blocks</code> can be used like this:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">f2_blocks[,,<span class="dv">1</span>]                <span class="co"># f2-statistics of the 1st SNP block</span></a>
<a class="sourceLine" id="cb9-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(f2_blocks, <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, mean)   <span class="co"># average across all blocks</span></a>
<a class="sourceLine" id="cb9-3" data-line-number="3">f2_blocks[pop1, pop2, ]       <span class="co"># f2(pop1, pop2) for all blocks</span></a></code></pre></div>
<p>The names along the 3rd dimension contain the SNP block lengths:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">block_lengths =<span class="st"> </span><span class="kw">parse_number</span>(<span class="kw"><a href="https://rdrr.io/r/base/dimnames.html">dimnames</a></span>(f2_blocks)[[<span class="dv">3</span>]])</a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(block_lengths)</a></code></pre></div>
<pre><code>## [1] 424 772 795 835 574 842</code></pre>
<p>If you want to try any of this without extracting and loading your own f2-statistics, you can instead use <code>example_f2_blocks</code> which becomes available as you load <code>admixtools</code>.</p>
<p><br></p>
<p>For more information on f-statistics in ADMIXTOOLS 2, click <a href="fstats.html">here</a>.</p>
<p><br></p>
</div>
<div id="qp3pop" class="section level2">
<h2 class="hasAnchor">
<a href="#qp3pop" class="anchor"></a>qp3Pop</h2>
<p>The original ADMIXTOOLS program for computing <span class="math inline">\(f_3\)</span>-statistics is called <em>qp3Pop</em>. In ADMIXTOOLS 2, you can compute <span class="math inline">\(f_3\)</span>-statistics like this:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">pop1 =<span class="st"> 'Denisova.DG'</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2">pop2 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'Altai_Neanderthal.DG'</span>, <span class="st">'Vindija.DG'</span>)</a>
<a class="sourceLine" id="cb12-3" data-line-number="3">pop3 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'Chimp.REF'</span>, <span class="st">'Mbuti.DG'</span>, <span class="st">'Russia_Ust_Ishim.DG'</span>)</a></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="kw"><a href="../reference/qp3pop.html">qp3pop</a></span>(f2_blocks, pop1, pop2, pop3)</a></code></pre></div>
<p>Or, equivalently</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw"><a href="../reference/qp3pop.html">f3</a></span>(f2_blocks, pop1, pop2, pop3)</a></code></pre></div>
<pre><code>## # A tibble: 6 x 7
##   pop1        pop2                 pop3                  est      se     z     p
##   &lt;chr&gt;       &lt;chr&gt;                &lt;chr&gt;               &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1 Denisova.DG Altai_Neanderthal.DG Chimp.REF          0.0591 5.98e-4  98.7     0
## 2 Denisova.DG Altai_Neanderthal.DG Mbuti.DG           0.0720 6.47e-4 111.      0
## 3 Denisova.DG Altai_Neanderthal.DG Russia_Ust_Ishim.… 0.0742 7.02e-4 106.      0
## 4 Denisova.DG Vindija.DG           Chimp.REF          0.0594 5.92e-4 100.      0
## 5 Denisova.DG Vindija.DG           Mbuti.DG           0.0724 6.34e-4 114.      0
## 6 Denisova.DG Vindija.DG           Russia_Ust_Ishim.… 0.0750 6.96e-4 108.      0</code></pre>
<p>This will compute f3-statistics for all combinations of <code>pop1</code>, <code>pop2</code>, and <code>pop3</code>. Typing just <code><a href="../reference/qp3pop.html">f3(f2_blocks)</a></code> will compute all possible combinations (which can be a large number). If only <code>pop1</code> is supplied, all combinations of populations in <code>pop1</code> will be computed.</p>
<p><br></p>
</div>
<div id="qpdstat" class="section level2">
<h2 class="hasAnchor">
<a href="#qpdstat" class="anchor"></a>qpDstat</h2>
<p>The original ADMIXTOOLS program for computing <span class="math inline">\(f_4\)</span>-statistics is called <em>qpDstat</em>. As the name suggests, it computes <em>D</em>-statistics by default. To get <span class="math inline">\(f_4\)</span>-statistics instead, the <code>f4mode</code> argument needs to set to <code>YES</code>. In ADMIXTOOLS 2, almost everything starts with f2-statistics, so the <code>qpdstat</code>/<code>f4</code> function computes <span class="math inline">\(f_4\)</span>-statistics by default.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">pop4 =<span class="st"> 'Switzerland_Bichon.SG'</span></a></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="kw"><a href="../reference/qpdstat.html">f4</a></span>(f2_blocks, pop1, pop2, pop3, pop4)</a>
<a class="sourceLine" id="cb17-2" data-line-number="2"><span class="kw"><a href="../reference/qpdstat.html">qpdstat</a></span>(f2_blocks, pop1, pop2, pop3, pop4)</a>
<a class="sourceLine" id="cb17-3" data-line-number="3"><span class="co"># two names for the same function</span></a></code></pre></div>
<pre><code>## # A tibble: 6 x 8
##   pop1     pop2        pop3        pop4             est      se      z         p
##   &lt;chr&gt;    &lt;chr&gt;       &lt;chr&gt;       &lt;chr&gt;          &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;     &lt;dbl&gt;
## 1 Denisov… Altai_Nean… Chimp.REF   Switzerlan…  1.50e-2 4.64e-4 32.3   6.06e-229
## 2 Denisov… Altai_Nean… Mbuti.DG    Switzerlan…  2.03e-3 3.53e-4  5.75  8.85e-  9
## 3 Denisov… Altai_Nean… Russia_Ust… Switzerlan… -2.17e-4 3.73e-4 -0.580 5.62e-  1
## 4 Denisov… Vindija.DG  Chimp.REF   Switzerlan…  1.54e-2 4.78e-4 32.2   5.81e-228
## 5 Denisov… Vindija.DG  Mbuti.DG    Switzerlan…  2.33e-3 3.63e-4  6.42  1.40e- 10
## 6 Denisov… Vindija.DG  Russia_Ust… Switzerlan… -2.40e-4 3.87e-4 -0.620 5.35e-  1</code></pre>
<p>The differences between <span class="math inline">\(f_4\)</span>-statistics and D-statistics are usually negligible. However, it is still possible to compute D-statistics in ADMIXTOOLS 2, by providing genotype data as the first argument, and setting <code>f4mode = FALSE</code>:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">prefix =<span class="st"> '/path/to/geno'</span></a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="kw"><a href="../reference/qpdstat.html">f4</a></span>(prefix, pop1, pop2, pop3, pop4, <span class="dt">f4mode =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p>Computing <span class="math inline">\(f_4\)</span>- or D-statistics from genotype data directly is slower, but it has the advantage that it avoids any problems that may arise from large amounts of missing data. More on this <a href="fstats.html">here</a>.</p>
<p><br></p>
</div>
<div id="qpwave-and-qpadm" class="section level2">
<h2 class="hasAnchor">
<a href="#qpwave-and-qpadm" class="anchor"></a>qpWave and qpAdm</h2>
<p><em>qpWave</em> and <em>qpAdm</em> are two programs with diffent goals - one estimates the number of admixture events, the other admixture weights - but they perform almost the same computations. The key difference is that <em>qpWave</em> is about comparing two sets of populations (<code>left</code> and <code>right</code>), while <em>qpAdm</em> is about testing how a single <code>target</code> population (which can be one of the <code>left</code> populations) relates to <code>left</code> and <code>right</code>. In ADMIXTOOLS 2, both <code><a href="../reference/qpadm.html">qpadm()</a></code> and <code><a href="../reference/qpwave.html">qpwave()</a></code>require at least three arguments:</p>
<ol style="list-style-type: decimal">
<li>
<span class="math inline">\(f_2\)</span>-statistics</li>
<li>A set of left populations</li>
<li>A set of right populations</li>
</ol>
<p><code><a href="../reference/qpadm.html">qpadm()</a></code> additionally requires a <code>target</code> population as the 4th argument, which will be modelled as a mixture of <code>left</code> populations.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1">left =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'Altai_Neanderthal.DG'</span>, <span class="st">'Vindija.DG'</span>)</a>
<a class="sourceLine" id="cb20-2" data-line-number="2">right =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'Chimp.REF'</span>, <span class="st">'Mbuti.DG'</span>, <span class="st">'Russia_Ust_Ishim.DG'</span>, <span class="st">'Switzerland_Bichon.SG'</span>)</a>
<a class="sourceLine" id="cb20-3" data-line-number="3">target =<span class="st"> 'Denisova.DG'</span></a>
<a class="sourceLine" id="cb20-4" data-line-number="4">pops =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(left, right, target)</a></code></pre></div>
<p>Both functions will return <span class="math inline">\(f_4\)</span>-statistics, and a data frame that shows how well the <span class="math inline">\(f_4\)</span>-matrix can be approximated by lower rank matrices. The last line tests for rank 0, which is equivalent to testing whether the left populations form a clade with respect to the right populations.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1">results =<span class="st"> </span><span class="kw"><a href="../reference/qpwave.html">qpwave</a></span>(f2_blocks, left, right)</a></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1">results<span class="op">$</span>f4</a></code></pre></div>
<pre><code>## # A tibble: 3 x 8
##   pop1           pop2     pop3     pop4                est      se     z       p
##   &lt;chr&gt;          &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt;             &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;
## 1 Altai_Neander… Vindija… Chimp.R… Mbuti.DG        1.24e-4 1.35e-4 0.920 0.358  
## 2 Altai_Neander… Vindija… Chimp.R… Russia_Ust_Ish… 4.45e-4 1.64e-4 2.72  0.00653
## 3 Altai_Neander… Vindija… Chimp.R… Switzerland_Bi… 4.22e-4 1.72e-4 2.45  0.0144</code></pre>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1">results<span class="op">$</span>rankdrop</a></code></pre></div>
<pre><code>## # A tibble: 1 x 7
##   f4rank   dof chisq       p dofdiff chisqdiff p_nested
## *  &lt;int&gt; &lt;int&gt; &lt;dbl&gt;   &lt;dbl&gt;   &lt;int&gt;     &lt;dbl&gt;    &lt;dbl&gt;
## 1      0     3  11.9 0.00768      NA        NA       NA</code></pre>
<p><code><a href="../reference/qpadm.html">qpadm()</a></code> will also compute admixture weights and nested models:</p>
<ul>
<li>
<strong>weights</strong>: These are the admixture weights, or estimates of the relative contributions of the left population to the target population.</li>
<li>
<strong>popdrop</strong>: popdrop shows the fits of all models generated by dropping a specific subset of left populations, and will only be returned if a target population is specified.</li>
<li>
<strong>f4</strong>: The estimated f4-statistics will now also include lines with fitted f4-statistics, where the target population is in the first column, and a weighted sum of the left populations, <code>fit</code>, in the second column.</li>
</ul>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1">results =<span class="st"> </span><span class="kw"><a href="../reference/qpadm.html">qpadm</a></span>(f2_blocks, left, right, target)</a></code></pre></div>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1">results<span class="op">$</span>weights</a></code></pre></div>
<pre><code>## # A tibble: 2 x 5
##   target      left                 weight    se     z
##   &lt;chr&gt;       &lt;chr&gt;                 &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1 Denisova.DG Altai_Neanderthal.DG   49.6  23.3  2.13
## 2 Denisova.DG Vindija.DG            -48.6  23.3 -2.08</code></pre>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1">results<span class="op">$</span>popdrop</a></code></pre></div>
<pre><code>## # A tibble: 3 x 13
##   pat      wt   dof  chisq      p f4rank Altai_Neanderth… Vindija.DG feasible
##   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;            &lt;dbl&gt;      &lt;dbl&gt; &lt;lgl&gt;   
## 1 00        0     2 7.15e0 0.0280      1             49.6      -48.6 FALSE   
## 2 01        1     3 1.14e4 0           0              1         NA   TRUE    
## 3 10        1     3 1.14e4 0           0             NA          1   TRUE    
## # … with 4 more variables: best &lt;lgl&gt;, dofdiff &lt;dbl&gt;, chisqdiff &lt;dbl&gt;,
## #   p_nested &lt;dbl&gt;</code></pre>
<p><br></p>
<p>The following two functions can be used to explore many <em>qpWave</em> or <em>qpAdm</em> models at the same time.</p>
<div id="pairwise-cladality-tests" class="section level4">
<h4 class="hasAnchor">
<a href="#pairwise-cladality-tests" class="anchor"></a>Pairwise cladality tests</h4>
<p><code><a href="../reference/qpwave_pairs.html">qpwave_pairs()</a></code> forms all pairs of <code>left</code> populations and tests whether they form a clade with respect to the <code>right</code> populations.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" data-line-number="1"><span class="kw"><a href="../reference/qpwave_pairs.html">qpwave_pairs</a></span>(f2_blocks, <span class="dt">left =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(target, left), <span class="dt">right =</span> right)</a></code></pre></div>
<pre><code>## # A tibble: 6 x 4
##   pop1                 pop2                  chisq       p
##   &lt;chr&gt;                &lt;chr&gt;                 &lt;dbl&gt;   &lt;dbl&gt;
## 1 Altai_Neanderthal.DG Denisova.DG          1507.  0      
## 2 Altai_Neanderthal.DG Vindija.DG             11.9 0.00768
## 3 Denisova.DG          Altai_Neanderthal.DG 1507.  0      
## 4 Denisova.DG          Vindija.DG           1510.  0      
## 5 Vindija.DG           Altai_Neanderthal.DG   11.9 0.00768
## 6 Vindija.DG           Denisova.DG          1510.  0</code></pre>
</div>
<div id="rotating-outgroups" class="section level4">
<h4 class="hasAnchor">
<a href="#rotating-outgroups" class="anchor"></a>Rotating outgroups</h4>
<p><code><a href="../reference/qpadm_rotate.html">qpadm_rotate()</a></code> tests many <code><a href="../reference/qpadm.html">qpadm()</a></code> models at a time. For each model, the <code>left</code> populations will be split into two groups: The first group will make up the actual <code>left</code> populations passed to <code><a href="../reference/qpadm.html">qpadm()</a></code>, while the second group will be added to the <code>right</code> populations.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" data-line-number="1"><span class="kw"><a href="../reference/qpadm_rotate.html">qpadm_rotate</a></span>(f2_blocks, <span class="dt">left =</span> pops[<span class="dv">3</span><span class="op">:</span><span class="dv">7</span>], <span class="dt">right =</span> pops[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>], <span class="dt">target =</span> pops[<span class="dv">1</span>])</a></code></pre></div>
<pre><code>## ℹ Evaluating 25 models...</code></pre>
<pre><code>## # A tibble: 25 x 7
##    left      right     f4rank   dof   chisq        p feasible
##    &lt;list&gt;    &lt;list&gt;     &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt; &lt;lgl&gt;   
##  1 &lt;chr [1]&gt; &lt;chr [4]&gt;      0     5 26194.  0.       TRUE    
##  2 &lt;chr [1]&gt; &lt;chr [4]&gt;      0     5 53965.  0.       TRUE    
##  3 &lt;chr [1]&gt; &lt;chr [4]&gt;      0     5 43564.  0.       TRUE    
##  4 &lt;chr [1]&gt; &lt;chr [4]&gt;      0     5 54909.  0.       TRUE    
##  5 &lt;chr [1]&gt; &lt;chr [4]&gt;      0     5 13167.  0.       TRUE    
##  6 &lt;chr [2]&gt; &lt;chr [3]&gt;      1     3 10602.  0.       FALSE   
##  7 &lt;chr [2]&gt; &lt;chr [3]&gt;      1     3 13506.  0.       FALSE   
##  8 &lt;chr [2]&gt; &lt;chr [3]&gt;      1     3 13447.  0.       FALSE   
##  9 &lt;chr [2]&gt; &lt;chr [3]&gt;      1     3    72.2 1.46e-15 FALSE   
## 10 &lt;chr [2]&gt; &lt;chr [3]&gt;      1     3  4030.  0.       FALSE   
## # … with 15 more rows</code></pre>
<p><br></p>
</div>
</div>
<div id="qpgraph" class="section level2">
<h2 class="hasAnchor">
<a href="#qpgraph" class="anchor"></a>qpGraph</h2>
<p>Single <span class="math inline">\(f_3\)</span>- and <span class="math inline">\(f_4\)</span>-statistics can tell us how three or four populations are related to each other. qpGraph generalizes this concept to any number of populations. It takes estimated <span class="math inline">\(f_3\)</span>-statistics and the topology of an admixtgure graph, finds the edges weights that minimize the difference between fitted and estimated <span class="math inline">\(f_3\)</span>-statistics, and summarizes that difference in a likelihood score.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1">qpg_results =<span class="st"> </span><span class="kw"><a href="../reference/qpgraph.html">qpgraph</a></span>(f2_dir, example_graph)</a>
<a class="sourceLine" id="cb36-2" data-line-number="2">qpg_results<span class="op">$</span>score</a></code></pre></div>
<pre><code>## [1] 19219.98</code></pre>
<p>Here, <code>example_graph</code> is a specific graph included in this R package, but you can provide any other graph in one of three formats.</p>
<ol style="list-style-type: decimal">
<li>An <a href="https://igraph.org/r/">igraph</a> object.</li>
<li>A two column matrix or data frame, where each row is an edge, the first column is the source of the edge, and the second column is the target. Additional columns labelled <code>lower</code> and <code>upper</code> can be used to constrain certain edges (<code>NA</code> = no constraint).</li>
<li>The location of a <em>qpGraph</em> graph file, which will be read and parsed.</li>
</ol>
<p>The leaf nodes of this graph have to match the <span class="math inline">\(f_2\)</span>-statistic population labels, and the graph has to be a valid admixture graph: a directed acyclic graph where each node has no more than two parents. If nodes with more than two children are present (polytomies or multifurcations) they will be split in a random order, and the new drift edges will be constrained to 0.</p>
<p><br></p>
<p>The output of <code>qpgraph</code> is a list with several items:</p>
<ul>
<li>
<strong>edges</strong>: A data frame with estimated edge weights. This includes normal edges as well as admixture edges.</li>
<li>
<strong>score</strong> The likelihood score. Smaller scores indicate a better fit.</li>
<li>
<strong>f2</strong> Estimated f2-statistics and standard errors for all population pairs</li>
<li>
<strong>f3</strong> Estimated and fitted f3-statistics for all population pairs with the outgroup. This includes residuals and Z-scores.</li>
<li>
<strong>opt</strong> A data frame with details from the weight optimization, with one row per set of starting values.</li>
<li>
<strong>ppinv</strong> The inverse of the f3-statistics covariance matrix</li>
</ul>
<p>Optionally, fitted and estimated f4-statistics can be returned by setting <code>return_f4 = TRUE</code>. When <code>f2_blocks_test</code> is provided, an out-of-sample score is computed and returned as <strong>score_test</strong>.</p>
<p>The fitted graph can be plotted like this:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" data-line-number="1"><span class="kw"><a href="../reference/plot_graph.html">plot_graph</a></span>(qpg_results<span class="op">$</span>edges)</a></code></pre></div>
<p><img src="admixtools_files/figure-html/unnamed-chunk-32-1.png" width="700"></p>
<p>or as an interactive plot:</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb39-1" data-line-number="1"><span class="kw"><a href="../reference/plotly_graph.html">plotly_graph</a></span>(qpg_results<span class="op">$</span>edges)</a></code></pre></div>
<div id="htmlwidget-453ce2174f838e45ff15" style="width:700px;height:432.632880098888px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-453ce2174f838e45ff15">{"x":{"data":[{"x":[0,2],"y":[-1,-2],"text":"N3N5 -> N3N8","type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(216,144,0,1)","dash":"dot"},"hoveron":"points","name":"(admix,-1)","legendgroup":"(admix,-1)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-2,1],"y":[0,-1],"text":"N2N3 -> N2N4","type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(163,165,0,1)","dash":"dot"},"hoveron":"points","name":"(admix,0)","legendgroup":"(admix,0)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-1.5,0],"y":[1,0],"text":"N3N2 -> N3N4","type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(57,182,0,1)","dash":"dot"},"hoveron":"points","name":"(admix,1)","legendgroup":"(admix,1)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0,0],"y":[2,0],"text":"N3N3 -> N3N4","type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(0,191,125,1)","dash":"dot"},"hoveron":"points","name":"(admix,2)","legendgroup":"(admix,2)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1,1,null,2,2],"y":[3,-1,null,3,-2],"text":["N2N2 -> N2N4","N2N2 -> N2N4",null,"N3N7 -> N3N8","N3N7 -> N3N8"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(0,191,196,1)","dash":"dot"},"hoveron":"points","name":"(admix,3)","legendgroup":"(admix,3)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[2,2],"y":[-2,-3],"text":"N3N8 -> Russia_Ust_Ishim.DG","type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(248,118,109,1)","dash":"solid"},"hoveron":"points","name":"(normal,-2)","legendgroup":"(normal,-2)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1,1],"y":[-1,-2],"text":"N2N4 -> Mbuti.DG","type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(216,144,0,1)","dash":"solid"},"hoveron":"points","name":"(normal,-1)","legendgroup":"(normal,-1)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0,0],"y":[0,-1],"text":"N3N4 -> N3N5","type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(163,165,0,1)","dash":"solid"},"hoveron":"points","name":"(normal,0)","legendgroup":"(normal,0)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-2.5,-3,null,-2.5,-2],"y":[1,0,null,1,0],"text":["N2N1 -> Vindija.DG","N2N1 -> Vindija.DG",null,"N2N1 -> N2N3","N2N1 -> N2N3"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(57,182,0,1)","dash":"solid"},"hoveron":"points","name":"(normal,1)","legendgroup":"(normal,1)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-2,-2.5,null,-2,-1.5],"y":[2,1,null,2,1],"text":["N3N0 -> N2N1","N3N0 -> N2N1",null,"N3N0 -> N3N2","N3N0 -> N3N2"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(0,191,125,1)","dash":"solid"},"hoveron":"points","name":"(normal,2)","legendgroup":"(normal,2)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-2.5,-3,null,-2.5,-2,null,-0.5,-1,null,-0.5,0],"y":[3,2,null,3,2,null,3,2,null,3,2],"text":["N0N -> Altai_Neanderthal.DG","N0N -> Altai_Neanderthal.DG",null,"N0N -> N3N0","N0N -> N3N0",null,"N3N1 -> Denisova.DG","N3N1 -> Denisova.DG",null,"N3N1 -> N3N3","N3N1 -> N3N3"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(0,191,196,1)","dash":"solid"},"hoveron":"points","name":"(normal,3)","legendgroup":"(normal,3)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-1.5,-2.5,null,-1.5,-0.5,null,1.5,1,null,1.5,2],"y":[4,3,null,4,3,null,4,3,null,4,3],"text":["N1N -> N0N","N1N -> N0N",null,"N1N -> N3N1","N1N -> N3N1",null,"N3N6 -> N2N2","N3N6 -> N2N2",null,"N3N6 -> N3N7","N3N6 -> N3N7"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(0,176,246,1)","dash":"solid"},"hoveron":"points","name":"(normal,4)","legendgroup":"(normal,4)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0,-1.5,null,0,1.5],"y":[5,4,null,5,4],"text":["N2N0 -> N1N","N2N0 -> N1N",null,"N2N0 -> N3N6","N2N0 -> N3N6"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(149,144,255,1)","dash":"solid"},"hoveron":"points","name":"(normal,5)","legendgroup":"(normal,5)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0.5,0,null,0.5,1],"y":[6,5,null,6,5],"text":["N4N -> N2N0","N4N -> N2N0",null,"N4N -> Switzerland_Bichon.SG","N4N -> Switzerland_Bichon.SG"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(231,107,243,1)","dash":"solid"},"hoveron":"points","name":"(normal,6)","legendgroup":"(normal,6)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0,-0.5,null,0,0.5],"y":[7,6,null,7,6],"text":["R -> Chimp.REF","R -> Chimp.REF",null,"R -> N4N","R -> N4N"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(255,98,188,1)","dash":"solid"},"hoveron":"points","name":"(normal,7)","legendgroup":"(normal,7)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-0.25,0.25,-2.75,-2.25,-2,-1,-0.75,0.75,-2.75,-2.25,1,-2.25,-1.75,-0.75,-0.25,0,1.25,1.75,2,0.25,0.75,1,-0.75,1,-0.5,0,2],"y":[6.5,6.5,2.5,2.5,3.5,3.5,4.5,4.5,0.5,0.5,-1.5,1.5,1.5,2.5,2.5,-0.5,3.5,3.5,-2.5,5.5,5.5,1,0.5,-1.5,-0.5,1,0.5],"text":["72","72","3","11","84","36","38","15","0","0","12","0","0","36","0","0","12","0","0","0","223","100%","100%","6%","0%","0%","94%"],"hovertext":["R -> Chimp.REF","R -> N4N","N0N -> Altai_Neanderthal.DG","N0N -> N3N0","N1N -> N0N","N1N -> N3N1","N2N0 -> N1N","N2N0 -> N3N6","N2N1 -> Vindija.DG","N2N1 -> N2N3","N2N4 -> Mbuti.DG","N3N0 -> N2N1","N3N0 -> N3N2","N3N1 -> Denisova.DG","N3N1 -> N3N3","N3N4 -> N3N5","N3N6 -> N2N2","N3N6 -> N3N7","N3N8 -> Russia_Ust_Ishim.DG","N4N -> N2N0","N4N -> Switzerland_Bichon.SG","N2N2 -> N2N4","N3N2 -> N3N4","N3N5 -> N3N8","N2N3 -> N2N4","N3N3 -> N3N4","N3N7 -> N3N8"],"textfont":{"size":9.4488188976378,"color":"rgba(0,0,0,1)"},"type":"scatter","mode":"text","hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[2],"y":[-3],"text":"Russia_Ust_Ishim.DG","hovertext":"","textfont":{"size":9.4488188976378,"color":"rgba(248,118,109,1)"},"type":"scatter","mode":"text","hoveron":"points","name":"(-2,1)","legendgroup":"(-2,1)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1],"y":[-2],"text":"Mbuti.DG","hovertext":"","textfont":{"size":9.4488188976378,"color":"rgba(216,144,0,1)"},"type":"scatter","mode":"text","hoveron":"points","name":"(-1,1)","legendgroup":"(-1,1)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-3],"y":[0],"text":"Vindija.DG","hovertext":"","textfont":{"size":9.4488188976378,"color":"rgba(57,182,0,1)"},"type":"scatter","mode":"text","hoveron":"points","name":"1","legendgroup":"1","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-3,-1],"y":[2,2],"text":["Altai_Neanderthal.DG","Denisova.DG"],"hovertext":["",""],"textfont":{"size":9.4488188976378,"color":"rgba(0,191,196,1)"},"type":"scatter","mode":"text","hoveron":"points","name":"(3,1)","legendgroup":"(3,1)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1],"y":[5],"text":"Switzerland_Bichon.SG","hovertext":"","textfont":{"size":9.4488188976378,"color":"rgba(231,107,243,1)"},"type":"scatter","mode":"text","hoveron":"points","name":"(6,1)","legendgroup":"(6,1)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-0.5],"y":[6],"text":"Chimp.REF","hovertext":"","textfont":{"size":9.4488188976378,"color":"rgba(255,98,188,1)"},"type":"scatter","mode":"text","hoveron":"points","name":"(7,1)","legendgroup":"(7,1)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0,-2.5,-1.5,0,-2.5,1,-2,-0.5,0,1.5,2,0.5,1,-1.5,0,-2,0,2],"y":[7,3,4,5,1,-1,2,3,0,4,-2,6,3,1,-1,0,2,3],"text":["R","N0N","N1N","N2N0","N2N1","N2N4","N3N0","N3N1","N3N4","N3N6","N3N8","N4N","N2N2","N3N2","N3N5","N2N3","N3N3","N3N7"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,0,0,1)","opacity":0,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":27.8383223406054,"r":7.30593607305936,"b":15.4912903771351,"l":10.958904109589},"paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[-3.6,2.6],"tickmode":"array","ticktext":["NA","-2","0","2"],"tickvals":[null,-2,4.44089209850063e-16,2],"categoryorder":"array","categoryarray":["NA","-2","0","2"],"nticks":null,"ticks":"","tickcolor":null,"ticklen":3.65296803652968,"tickwidth":0,"showticklabels":false,"tickfont":{"color":null,"family":null,"size":0},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(255,255,255,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"y","title":{"text":"","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[-3.5,7.5],"tickmode":"array","ticktext":["-2.5","0.0","2.5","5.0","7.5"],"tickvals":[-2.5,0,2.5,5,7.5],"categoryorder":"array","categoryarray":["-2.5","0.0","2.5","5.0","7.5"],"nticks":null,"ticks":"","tickcolor":null,"ticklen":3.65296803652968,"tickwidth":0,"showticklabels":false,"tickfont":{"color":null,"family":null,"size":0},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(255,255,255,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"x","title":{"text":"","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":false,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895}},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","showSendToCloud":false},"source":"A","attrs":{"14ab17c4aad49":{"linetype":{},"colour":{},"text":{},"x":{},"y":{},"xend":{},"yend":{},"from":{},"to":{},"type":"scatter"},"14ab1ab6d35f":{"x":{},"y":{},"label":{},"text":{},"x.1":{},"y.1":{},"xend":{},"yend":{},"from":{},"to":{}},"14ab168ef0777":{"label":{},"colour":{},"from":{},"x":{},"y":{},"xend":{},"yend":{},"from.1":{},"to":{}},"14ab124d34b73":{"x":{},"y":{},"text":{},"x.1":{},"y.1":{},"xend":{},"yend":{},"from":{},"to":{}}},"cur_data":"14ab17c4aad49","visdat":{"14ab17c4aad49":["function (y) ","x"],"14ab1ab6d35f":["function (y) ","x"],"14ab168ef0777":["function (y) ","x"],"14ab124d34b73":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script><p><br></p>
<div id="more-about-graphs" class="section level3">
<h3 class="hasAnchor">
<a href="#more-about-graphs" class="anchor"></a>More about graphs</h3>
<p>ADMIXTOOLS 2 can automatically find well-fitting graphs, and it has several functions for exploring the topological neighborhood of a graph. These functions are described <a href="graphs.html">here</a>.</p>
<p><br></p>
</div>
</div>
<div id="comparing-results-to-admixtools" class="section level2">
<h2 class="hasAnchor">
<a href="#comparing-results-to-admixtools" class="anchor"></a>Comparing results to ADMIXTOOLS</h2>
<p>There are wrapper function which call the origial ADMIXTOOLS programs and read the results. This can be used to check if the results of <code>qpgraph</code> are correct.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" data-line-number="1">binpath =<span class="st"> '/home/np29/o2bin/'</span></a>
<a class="sourceLine" id="cb40-2" data-line-number="2">env =<span class="st"> 'export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/n/app/openblas/0.2.19/lib/:/n/app/gsl/2.3/lib/;'</span></a>
<a class="sourceLine" id="cb40-3" data-line-number="3"></a>
<a class="sourceLine" id="cb40-4" data-line-number="4">qp3pop_bin  =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(env, binpath, <span class="st">'qp3pop'</span>)</a>
<a class="sourceLine" id="cb40-5" data-line-number="5">qpdstat_bin =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(env, binpath, <span class="st">'qpDstat'</span>)</a>
<a class="sourceLine" id="cb40-6" data-line-number="6">qpadm_bin   =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(env, binpath, <span class="st">'qpAdm'</span>)</a>
<a class="sourceLine" id="cb40-7" data-line-number="7">qpgraph_bin =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(env, binpath, <span class="st">'qpGraph'</span>)</a>
<a class="sourceLine" id="cb40-8" data-line-number="8"></a>
<a class="sourceLine" id="cb40-9" data-line-number="9"><span class="co">#prefix = '/n/groups/reich/DAVID/V42/V42.1/v42.1'</span></a>
<a class="sourceLine" id="cb40-10" data-line-number="10">prefix =<span class="st"> '/n/groups/reich/robert/projects/admixprograms/v42.1_small'</span></a>
<a class="sourceLine" id="cb40-11" data-line-number="11">outdir =<span class="st"> 'write/files/here/'</span></a>
<a class="sourceLine" id="cb40-12" data-line-number="12"></a>
<a class="sourceLine" id="cb40-13" data-line-number="13"><span class="kw"><a href="../reference/qp3pop_wrapper.html">qp3pop_wrapper</a></span>(<span class="dt">source1 =</span> pop2, <span class="dt">source2 =</span> pop3, <span class="dt">target =</span> pop1,</a>
<a class="sourceLine" id="cb40-14" data-line-number="14">               qp3pop_bin, prefix, outdir)</a>
<a class="sourceLine" id="cb40-15" data-line-number="15"><span class="kw"><a href="../reference/qpdstat_wrapper.html">qpdstat_wrapper</a></span>(pop1, pop2, pop3, pop4,</a>
<a class="sourceLine" id="cb40-16" data-line-number="16">                qpdstat_bin, prefix, outdir)</a>
<a class="sourceLine" id="cb40-17" data-line-number="17"><span class="kw"><a href="../reference/qpadm_wrapper.html">qpadm_wrapper</a></span>(left, right, target,</a>
<a class="sourceLine" id="cb40-18" data-line-number="18">              qpadm_bin, prefix, outdir)</a>
<a class="sourceLine" id="cb40-19" data-line-number="19"><span class="kw"><a href="../reference/qpgraph_wrapper.html">qpgraph_wrapper</a></span>(example_graph,</a>
<a class="sourceLine" id="cb40-20" data-line-number="20">                qpgraph_bin, prefix, outdir)</a></code></pre></div>
<p><strong>Unless <code>outdir</code> is specified, calling these wrapper functions may overwrite files in the working directory!</strong></p>
<p>If you already have existing parameter files and population or graph files, you can run the wrapper functions like this (though the different programs will require different parfiles and popfiles):</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb41-1" data-line-number="1"><span class="kw"><a href="../reference/qp3pop_wrapper.html">qp3pop_wrapper</a></span> (<span class="dt">parfile =</span> <span class="st">'parfile.txt'</span>, <span class="dt">bin =</span> qp3pop_bin)</a>
<a class="sourceLine" id="cb41-2" data-line-number="2"><span class="kw"><a href="../reference/qpdstat_wrapper.html">qpdstat_wrapper</a></span>(<span class="dt">parfile =</span> <span class="st">'parfile.txt'</span>, <span class="dt">bin =</span> qpdstat_bin)</a>
<a class="sourceLine" id="cb41-3" data-line-number="3"><span class="kw"><a href="../reference/qpadm_wrapper.html">qpadm_wrapper</a></span>  (<span class="dt">parfile =</span> <span class="st">'parfile.txt'</span>, <span class="dt">bin =</span> qpadm_bin)</a>
<a class="sourceLine" id="cb41-4" data-line-number="4"></a>
<a class="sourceLine" id="cb41-5" data-line-number="5"><span class="kw"><a href="../reference/qp3pop_wrapper.html">qp3pop_wrapper</a></span> (<span class="st">'popfile.txt'</span>, <span class="dt">bin =</span> qp3pop_bin,  <span class="dt">pref =</span> prefix)</a>
<a class="sourceLine" id="cb41-6" data-line-number="6"><span class="kw"><a href="../reference/qpdstat_wrapper.html">qpdstat_wrapper</a></span>(<span class="st">'popfile.txt'</span>, <span class="dt">bin =</span> qpdstat_bin, <span class="dt">pref =</span> prefix)</a>
<a class="sourceLine" id="cb41-7" data-line-number="7"></a>
<a class="sourceLine" id="cb41-8" data-line-number="8"><span class="kw"><a href="../reference/qpadm_wrapper.html">qpadm_wrapper</a></span>(<span class="dt">left =</span> <span class="st">'left.txt'</span>, <span class="dt">right =</span> <span class="st">'right.txt'</span>, <span class="dt">bin =</span> qpadm_bin, <span class="dt">pref =</span> prefix)</a>
<a class="sourceLine" id="cb41-9" data-line-number="9"></a>
<a class="sourceLine" id="cb41-10" data-line-number="10"><span class="kw"><a href="../reference/qpgraph_wrapper.html">qpgraph_wrapper</a></span>(<span class="st">'graphfile.txt'</span>, <span class="dt">parfile =</span> <span class="st">'parfile.txt'</span>, <span class="dt">bin =</span> qpgraph_bin)</a></code></pre></div>
<p><br></p>
<p>The following function makes it easy to compare qpGraph or qpAdm results.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" data-line-number="1">qpgraph_ref_results =<span class="st"> </span><span class="kw"><a href="../reference/qpgraph_wrapper.html">qpgraph_wrapper</a></span>(example_graph, qpgraph_bin, prefix)</a></code></pre></div>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb43-1" data-line-number="1"><span class="kw"><a href="../reference/plot_comparison.html">plot_comparison</a></span>(qpg_results, qpgraph_ref_results)</a></code></pre></div>
<p><img src="admixtools_files/figure-html/unnamed-chunk-38-1.png" width="700"><br></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#f-statistics"><em>f</em>-statistics</a></li>
      <li><a href="#qp3pop">qp3Pop</a></li>
      <li><a href="#qpdstat">qpDstat</a></li>
      <li><a href="#qpwave-and-qpadm">qpWave and qpAdm</a></li>
      <li><a href="#qpgraph">qpGraph</a></li>
      <li><a href="#comparing-results-to-admixtools">Comparing results to ADMIXTOOLS</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Robert Maier, Nick Patterson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '',
    indexName: 'admixtools',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
