<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>qpWave and qpAdm • admixtools</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><script src="../extra.js"></script><meta property="og:title" content="qpWave and qpAdm">
<meta property="og:description" content="admixtools">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-155330476-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-155330476-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">admixtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/admixtools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/fstats.html">f-statistics</a>
    </li>
    <li>
      <a href="../articles/graphs.html">Admixture graphs</a>
    </li>
    <li>
      <a href="../articles/io.html">Data formats</a>
    </li>
    <li>
      <a href="../articles/parallel.html">Parallelization</a>
    </li>
    <li>
      <a href="../articles/plotting.html">Plotting</a>
    </li>
    <li>
      <a href="../articles/qpadm.html">qpWave and qpAdm</a>
    </li>
    <li>
      <a href="../articles/resampling.html">Standard errors</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/uqrmaie1/admixtools/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="qpadm_files/header-attrs-2.5/header-attrs.js"></script><link href="qpadm_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="qpadm_files/anchor-sections-1.0/anchor-sections.js"></script><script src="qpadm_files/htmlwidgets-1.5.2/htmlwidgets.js"></script><script src="qpadm_files/plotly-binding-4.9.2.1/plotly.js"></script><script src="qpadm_files/typedarray-0.1/typedarray.min.js"></script><script src="qpadm_files/jquery-1.11.3/jquery.min.js"></script><link href="qpadm_files/crosstalk-1.1.0.1/css/crosstalk.css" rel="stylesheet">
<script src="qpadm_files/crosstalk-1.1.0.1/js/crosstalk.min.js"></script><link href="qpadm_files/plotly-htmlwidgets-css-1.52.2/plotly-htmlwidgets.css" rel="stylesheet">
<script src="qpadm_files/plotly-main-1.52.2/plotly-latest.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>qpWave and qpAdm</h1>
                        <h4 class="author">Robert Maier</h4>
            
            <h4 class="date">2021-01-26</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/uqrmaie1/admixtools/blob/master/vignettes/qpadm.Rmd"><code>vignettes/qpadm.Rmd</code></a></small>
      <div class="hidden name"><code>qpadm.Rmd</code></div>

    </div>

    
    
<p><strong>Under construction</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org">magrittr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/uqrmaie1/admixtools">admixtools</a></span><span class="op">)</span></code></pre></div>
<p><em>qpWave</em> and <em>qpAdm</em> are programs which can answer a number of questions about the relationships of different populations, such as:</p>
<ul>
<li>Do populations form clades with respect to each other?</li>
<li>What is the minimum number of admixture waves that separates two groups of populations?</li>
<li>Can a population be modeled as a mix of a set of other populations, or would this miss an ancestry component?</li>
<li>What are the relative proportions of the different admixing populations?</li>
</ul>
<p>The first two questions can be answered by <em>qpWave</em>. The latter two questions are about a specific target population, and they can be answered by <em>qpAdm</em>, an extension of <em>qpWave</em>.</p>
<p>This page describes how both programs use <span class="math inline">\(f_4\)</span>-statistics to address these questions. The original description of <em>qpWave</em> and <em>qpAdm</em> which goes into more technical detail can be found in Supplementary Information section 10 in <a href="https://www.nature.com/articles/nature14317">this paper</a>.</p>
<div id="qpwave" class="section level2">
<h2 class="hasAnchor">
<a href="#qpwave" class="anchor"></a>qpWave</h2>
<div id="qpwave-and-f4" class="section level3">
<h3 class="hasAnchor">
<a href="#qpwave-and-f4" class="anchor"></a>qpWave and f4</h3>
<p><span class="math inline">\(f_4\)</span>-statistics can be used to test whether two pairs of populations (<span class="math inline">\(L_1, L_2\)</span> and <span class="math inline">\(R_1, R_2\)</span>) form clades with respect to one another: If <span class="math inline">\(f_4(L_1,L_2; R_1,R_2)\)</span> is significantly different from 0, the two groups do not form a clade. If we have two groups with more than two populations each, and we want to know if the groups form two clades, we could compute all possible <span class="math inline">\(f_4\)</span>-statistics and test if any of them deviate from zero. Except we would have to correct for multiple testing, which is not straight-forward in this case because the <span class="math inline">\(f_4\)</span>-statistics are not independent of each other.</p>
<p><em>qpWave</em> provides another way to test whether two groups of populations form two clades. We can empirically see that <em>qpWave</em> and <span class="math inline">\(f_4\)</span> are closely related, because in the special case of applying <em>qpWave</em> to only four populations, we get almost the same p-value as for an <span class="math inline">\(f_4\)</span>-statistic of the four populations:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">L</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Altai_Neanderthal.DG'</span>, <span class="st">'Vindija.DG'</span><span class="op">)</span>
<span class="va">R</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Chimp.REF'</span>, <span class="st">'Mbuti.DG'</span><span class="op">)</span>
<span class="fu"><a href="../reference/qpdstat.html">f4</a></span><span class="op">(</span><span class="va">example_f2_blocks</span>, <span class="va">L</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">L</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="va">R</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">R</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">p</span></code></pre></div>
<pre><code>## [1] 0.3576872</code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/qpwave.html">qpwave</a></span><span class="op">(</span><span class="va">example_f2_blocks</span>, <span class="va">L</span>, <span class="va">R</span><span class="op">)</span><span class="op">$</span><span class="va">rankdrop</span><span class="op">$</span><span class="va">p</span></code></pre></div>
<pre><code>## ℹ Computing f4 stats...
## ℹ Computing number of admixture waves...</code></pre>
<pre><code>## [1] 0.3577113</code></pre>
</div>
<div id="estimating-the-number-of-independent-gene-flows" class="section level3">
<h3 class="hasAnchor">
<a href="#estimating-the-number-of-independent-gene-flows" class="anchor"></a>Estimating the number of independent gene flows</h3>
<p>Testing whether two pairs of populations form clades with each other is the same as asking whether there is only one gene flow event which separates the two pairs (<span class="math inline">\(f_4 = 0\)</span>) or whether there must have been more than a single gene flow event between the two pairs (<span class="math inline">\(f_4 != 0\)</span>).</p>
<p>In the graph on the left, <span class="math inline">\(L1\)</span> and <span class="math inline">\(L2\)</span> form a clade relative to <span class="math inline">\(R1\)</span> and <span class="math inline">\(R2\)</span>, which means <span class="math inline">\(L\)</span> and <span class="math inline">\(R\)</span> are only connected through one gene flow event. On the right they do not form a clade, and they are connected through two independent gene flow events (one going through the root, and the other going through the edge <code>n1b|n2b</code>).</p>
<div style="display: flex">
<div style="width: 50%">
<div id="htmlwidget-1cdcae464ec8abe005cc" style="width:100%;height:432.632880098888px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-1cdcae464ec8abe005cc">{"x":{"data":[{"x":[-1,-1.5,null,-1,-0.5,null,1,0.5,null,1,1.5],"y":[1,0,null,1,0,null,1,0,null,1,0],"text":["n1 -> L1","n1 -> L1",null,"n1 -> L2","n1 -> L2",null,"n2 -> R1","n2 -> R1",null,"n2 -> R2","n2 -> R2"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(248,118,109,1)","dash":"solid"},"hoveron":"points","name":"(normal,1)","legendgroup":"(normal,1)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0,-1,null,0,1],"y":[2,1,null,2,1],"text":["R -> n1","R -> n1",null,"R -> n2","R -> n2"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(0,191,196,1)","dash":"solid"},"hoveron":"points","name":"(normal,2)","legendgroup":"(normal,2)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0,-1,1],"y":[2,1,1],"text":["R","n1","n2"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,0,0,1)","opacity":0,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-0.5,0.5,-1.25,-0.75,0.75,1.25],"y":[1.5,1.5,0.5,0.5,0.5,0.5],"text":["","","","","",""],"hovertext":["R -> n1","R -> n2","n1 -> L1","n1 -> L2","n2 -> R1","n2 -> R2"],"textfont":{"size":9.4488188976378,"color":"rgba(0,0,0,1)"},"type":"scatter","mode":"text","hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-1.5,-0.5,0.5,1.5],"y":[-0.1,-0.1,-0.1,-0.1],"text":["L1","L2","R1","R2"],"hovertext":["","","",""],"textfont":{"size":9.4488188976378,"color":"rgba(248,118,109,1)"},"type":"scatter","mode":"text","hoveron":"points","name":"1","legendgroup":"1","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":27.8246364188244,"r":7.30593607305936,"b":15.4776044553541,"l":10.958904109589},"paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[-1.9,1.9],"tickmode":"array","ticktext":["-1","0","1"],"tickvals":[-1,0,1],"categoryorder":"array","categoryarray":["-1","0","1"],"nticks":null,"ticks":"","tickcolor":null,"ticklen":3.65296803652968,"tickwidth":0,"showticklabels":false,"tickfont":{"color":null,"family":null,"size":0},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(255,255,255,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"y","title":{"text":"","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[-0.205,2.105],"tickmode":"array","ticktext":["0.0","0.5","1.0","1.5","2.0"],"tickvals":[0,0.5,1,1.5,2],"categoryorder":"array","categoryarray":["0.0","0.5","1.0","1.5","2.0"],"nticks":null,"ticks":"","tickcolor":null,"ticklen":3.65296803652968,"tickwidth":0,"showticklabels":false,"tickfont":{"color":null,"family":null,"size":0},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(255,255,255,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"x","title":{"text":"","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":false,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895}},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","showSendToCloud":false},"source":"A","attrs":{"100e374c8730a":{"linetype":{},"colour":{},"text":{},"x":{},"y":{},"xend":{},"yend":{},"from":{},"to":{},"type":"scatter"},"100e313f8d16c":{"x":{},"y":{},"text":{},"x.1":{},"y.1":{},"xend":{},"yend":{},"from":{},"to":{}},"100e3347d13b2":{"x":{},"y":{},"label":{},"text":{},"x.1":{},"y.1":{},"xend":{},"yend":{},"from":{},"to":{}},"100e37f982609":{"label":{},"colour":{},"from":{},"x":{},"y":{},"xend":{},"yend":{},"from.1":{},"to":{}}},"cur_data":"100e374c8730a","visdat":{"100e374c8730a":["function (y) ","x"],"100e313f8d16c":["function (y) ","x"],"100e3347d13b2":["function (y) ","x"],"100e37f982609":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
</div>
<div style="width: 50%">
<div id="htmlwidget-0b881e941d96f63d523a" style="width:100%;height:432.632880098888px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-0b881e941d96f63d523a">{"x":{"data":[{"x":[-1.5,0.5],"y":[1,0],"text":"n1b -> n2b","type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(124,174,0,1)","dash":"dot"},"hoveron":"points","name":"(admix,1)","legendgroup":"(admix,1)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1,0.5],"y":[2,0],"text":"n2 -> n2b","type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(0,191,196,1)","dash":"dot"},"hoveron":"points","name":"(admix,2)","legendgroup":"(admix,2)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0.5,0.5],"y":[0,-1],"text":"n2b -> R1","type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(248,118,109,1)","dash":"solid"},"hoveron":"points","name":"(normal,0)","legendgroup":"(normal,0)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-1.5,-1.5],"y":[1,0],"text":"n1b -> L2","type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(124,174,0,1)","dash":"solid"},"hoveron":"points","name":"(normal,1)","legendgroup":"(normal,1)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-1,-1.5,null,-1,-0.5,null,1,1.5],"y":[2,1,null,2,1,null,2,1],"text":["n1 -> n1b","n1 -> n1b",null,"n1 -> L1","n1 -> L1",null,"n2 -> R2","n2 -> R2"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(0,191,196,1)","dash":"solid"},"hoveron":"points","name":"(normal,2)","legendgroup":"(normal,2)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0,-1,null,0,1],"y":[3,2,null,3,2],"text":["R -> n1","R -> n1",null,"R -> n2","R -> n2"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(199,124,255,1)","dash":"solid"},"hoveron":"points","name":"(normal,3)","legendgroup":"(normal,3)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0,-1,-1.5,0.5,1],"y":[3,2,1,0,2],"text":["R","n1","n1b","n2b","n2"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,0,0,1)","opacity":0,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-0.5,0.5,-1.25,-1.5,-0.5,-0.75,0.5,1.25,0.75],"y":[2.5,2.5,1.5,0.5,0.5,1.5,-0.5,1.5,1],"text":["","","","","","","","",""],"hovertext":["R -> n1","R -> n2","n1 -> n1b","n1b -> L2","n1b -> n2b","n1 -> L1","n2b -> R1","n2 -> R2","n2 -> n2b"],"textfont":{"size":9.4488188976378,"color":"rgba(0,0,0,1)"},"type":"scatter","mode":"text","hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0.5],"y":[-1.1],"text":"R1","hovertext":"","textfont":{"size":9.4488188976378,"color":"rgba(248,118,109,1)"},"type":"scatter","mode":"text","hoveron":"points","name":"(0,1)","legendgroup":"(0,1)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-1.5],"y":[-0.1],"text":"L2","hovertext":"","textfont":{"size":9.4488188976378,"color":"rgba(124,174,0,1)"},"type":"scatter","mode":"text","hoveron":"points","name":"1","legendgroup":"1","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-0.5,1.5],"y":[0.9,0.9],"text":["L1","R2"],"hovertext":["",""],"textfont":{"size":9.4488188976378,"color":"rgba(0,191,196,1)"},"type":"scatter","mode":"text","hoveron":"points","name":"(2,1)","legendgroup":"(2,1)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":27.8246364188244,"r":7.30593607305936,"b":15.4776044553541,"l":10.958904109589},"paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[-1.9,1.9],"tickmode":"array","ticktext":["-1","0","1"],"tickvals":[-1,0,1],"categoryorder":"array","categoryarray":["-1","0","1"],"nticks":null,"ticks":"","tickcolor":null,"ticklen":3.65296803652968,"tickwidth":0,"showticklabels":false,"tickfont":{"color":null,"family":null,"size":0},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(255,255,255,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"y","title":{"text":"","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[-1.305,3.205],"tickmode":"array","ticktext":["-1","0","1","2","3"],"tickvals":[-1,0,1,2,3],"categoryorder":"array","categoryarray":["-1","0","1","2","3"],"nticks":null,"ticks":"","tickcolor":null,"ticklen":3.65296803652968,"tickwidth":0,"showticklabels":false,"tickfont":{"color":null,"family":null,"size":0},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(255,255,255,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"x","title":{"text":"","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":false,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895}},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","showSendToCloud":false},"source":"A","attrs":{"100e37a6c1b6":{"linetype":{},"colour":{},"text":{},"x":{},"y":{},"xend":{},"yend":{},"from":{},"to":{},"type":"scatter"},"100e354f797a6":{"x":{},"y":{},"text":{},"x.1":{},"y.1":{},"xend":{},"yend":{},"from":{},"to":{}},"100e34b053ede":{"x":{},"y":{},"label":{},"text":{},"x.1":{},"y.1":{},"xend":{},"yend":{},"from":{},"to":{}},"100e34562874c":{"label":{},"colour":{},"from":{},"x":{},"y":{},"xend":{},"yend":{},"from.1":{},"to":{}}},"cur_data":"100e37a6c1b6","visdat":{"100e37a6c1b6":["function (y) ","x"],"100e354f797a6":["function (y) ","x"],"100e34b053ede":["function (y) ","x"],"100e34562874c":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<p><em>qpWave</em> allows us to estimate a lower bound on the number of gene flow events that separate two groups of populations <span class="math inline">\(L\)</span> and <span class="math inline">\(R\)</span> of size <span class="math inline">\(n_L\)</span> and <span class="math inline">\(n_R\)</span>.</p>
<p>The basic idea behind it is that we can use <span class="math inline">\(f_4\)</span>-statistics of the form <span class="math inline">\(f_4(L_i, L_j; R_m, R_n)\)</span> to estimate the genetic drift that is shared between pairs of left populations and pairs of right populations. There are <span class="math inline">\(\frac{n_L(n_L-1)}{2} \frac{n_R(n_R-1)}{2}\)</span> <span class="math inline">\(f_4\)</span>-statistics of that form, but not all of them contain unique information. In the case where <span class="math inline">\(L\)</span> and <span class="math inline">\(R\)</span> form two clades, all <span class="math inline">\(f_4\)</span>-statistics <span class="math inline">\(f_4(L_i, L_j; R_m, R_n)\)</span> are expected to be zero. If there are two independent gene flow events, a few of these <span class="math inline">\(f_4\)</span>-statistics will be enough to determine what the other <span class="math inline">\(f_4\)</span>-statistics should be. More generally, we can construct a matrix <span class="math inline">\(X\)</span> of these <span class="math inline">\(f_4\)</span>-statistics, and the rank of <span class="math inline">\(X\)</span> tells us what the minimum number of gene flows is between <span class="math inline">\(L\)</span> and <span class="math inline">\(R\)</span>. A high rank of <span class="math inline">\(X\)</span> suggests many independent gene flows between <span class="math inline">\(L\)</span> and <span class="math inline">\(R\)</span>.</p>
<p>In practice, <span class="math inline">\(X\)</span> is constructed by fixing one population in <span class="math inline">\(L\)</span> and one population in <span class="math inline">\(R\)</span> which results in a matrix of dimensions <span class="math inline">\((n_L - 1) \times (n_R - 1)\)</span> with all <span class="math inline">\(f_4\)</span>-statistics of the form <span class="math inline">\(f_4(L_1, L_i; R_1, R_j)\)</span>. By convention, <span class="math inline">\(n_R &gt; n_L\)</span>, which means that the highest possible rank of <span class="math inline">\(X\)</span> is <span class="math inline">\(n_L - 1\)</span>. A rank of <span class="math inline">\(n_L - 1\)</span> suggests that at least <span class="math inline">\(n_L - 1\)</span> independent gene flows have occurred between <span class="math inline">\(L\)</span> and <span class="math inline">\(R\)</span>.</p>
</div>
<div id="estimating-the-rank-of-x" class="section level3">
<h3 class="hasAnchor">
<a href="#estimating-the-rank-of-x" class="anchor"></a>Estimating the rank of X</h3>
<p>To estimate the rank of <span class="math inline">\(X\)</span> we test how well <span class="math inline">\(X\)</span> can be approximated by different low rank approximations. An approximation of <span class="math inline">\(X\)</span> with one rank less than full rank can be written as <span class="math inline">\(AB\)</span>, where <span class="math inline">\(A\)</span> is <span class="math inline">\((n_L-1) \times (n_L-2)\)</span>, and <span class="math inline">\(B\)</span> is <span class="math inline">\((n_L-2) \times (n_R-1)\)</span> and <span class="math inline">\(AB\)</span> should be as close as possible to <span class="math inline">\(X\)</span>. In this case (one rank less than full rank), <span class="math inline">\(X\)</span> would have rank <span class="math inline">\(n_L - 1\)</span>, and <span class="math inline">\(AB\)</span> would have rank <span class="math inline">\(n_L - 2\)</span>. If this approximation works well, the residuals <span class="math inline">\(E = X - AB\)</span> will be very small. Large residuals suggest that the low rank approximation <span class="math inline">\(AB\)</span> doesn’t capture all the information in <span class="math inline">\(X\)</span> and that a simple model with only few admixture events doesn’t fit the data. How large these residuals can become before a model can be rejected depends on the standard errors of the <span class="math inline">\(f_4\)</span>-statistics. If we have data from a large number of SNPs, these standard errors will be small, which gives us more power to reject models.</p>
<p>To get low rank approximations of <span class="math inline">\(X\)</span>, we start with a singular value decompositions (SVD) of <span class="math inline">\(X\)</span>, which gives us precursors of the matrices <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span>. Because we cannot ignore the covariance among the <span class="math inline">\(f_4\)</span>-statistics (<span class="math inline">\(Q\)</span>), <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span> need to be further adjusted. The initial SVD estimates of <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span> already minimize the squared sum of the residuals <span class="math inline">\(E = X - AB\)</span>. What is actually minimized in the end is the quadratic form <span class="math inline">\(E' Q^{-1} E\)</span> (<span class="math inline">\(E\)</span> is a vector of length <span class="math inline">\((n_L - 1)(n_R - 1)\)</span>, and <span class="math inline">\(Q\)</span> is a square matrix of dimension <span class="math inline">\((n_L - 1)(n_R - 1) \times (n_L - 1)(n_R - 1)\)</span>). This quantity is closely related to the likelihood of the model, <span class="math inline">\(L(A,B) = -\frac{1}{2} E'Q^{-1}E\)</span>. The difference between the log likelihoods of two models (same populations, but one with higher rank than the other) approximately follows a <span class="math inline">\(\chi^2\)</span> distribution with <span class="math inline">\(x\)</span> degrees of freedom. From this <span class="math inline">\(\chi^2\)</span> distribution we can compute a p-value that tests whether we can reject simple models with only one or a few admixture waves.</p>
<p><br></p>
<p>Let’s consider two extreme cases:</p>
<p>First, <span class="math inline">\(L\)</span> and <span class="math inline">\(R\)</span> are two groups of populations which form clades with respect to each other. This means that all <span class="math inline">\(f_4\)</span>-statistics of the form <span class="math inline">\(f_4(L_i, L_j; R_m, R_n)\)</span> are <span class="math inline">\(0\)</span>, and hence <span class="math inline">\(X\)</span> is just a matrix full of <span class="math inline">\(0s\)</span> and rank has rank <span class="math inline">\(0\)</span>.</p>
<p>The opposite extreme is when <span class="math inline">\(X\)</span> (which has dimensions <span class="math inline">\((n_L - 1) \times (n_R - 1)\)</span>) has full rank (<span class="math inline">\((n_L - 1)\)</span>, because <span class="math inline">\(n_L\)</span> &lt;= <span class="math inline">\(n_R\)</span>). This means that there is no population in <span class="math inline">\(L\)</span> which can be modeled as a combination of other populations in <span class="math inline">\(L\)</span>. Each population in <span class="math inline">\(L\)</span> shares some amount of drift with populations in <span class="math inline">\(R\)</span>, which is not captured by other populations in <span class="math inline">\(L\)</span>.</p>
<p>This second extreme case is important, because it forms the basis for <em>qpAdm</em>: Once we have identified two groups <span class="math inline">\(L\)</span> and <span class="math inline">\(R\)</span> with this property, we can add an additional population <span class="math inline">\(T\)</span> and test whether this property is retained or not.</p>
</div>
</div>
<div id="qpadm" class="section level2">
<h2 class="hasAnchor">
<a href="#qpadm" class="anchor"></a>qpAdm</h2>
<div id="modelling-an-admixed-population" class="section level3">
<h3 class="hasAnchor">
<a href="#modelling-an-admixed-population" class="anchor"></a>Modelling an admixed population</h3>
<p>The <em>qpWave</em> section above describes how we can use a matrix of <span class="math inline">\(f_4\)</span>-statistics (<span class="math inline">\(X\)</span>) to test the number of gene flows that separate <span class="math inline">\(L\)</span> and <span class="math inline">\(R\)</span>. The idea behind <em>qpAdm</em> is that we can say something about an additional population <span class="math inline">\(T\)</span> if we follow this approach twice: Once for a model with <span class="math inline">\(L\)</span> and <span class="math inline">\(R\)</span>, and once for a model with <span class="math inline">\(L \cup T\)</span> (the union of <span class="math inline">\(L\)</span> and <span class="math inline">\(T\)</span>) and <span class="math inline">\(R\)</span>. If we find a higher number of gene flows for the model with <span class="math inline">\(L \cup T\)</span> and <span class="math inline">\(R\)</span> than for the model with only <span class="math inline">\(L\)</span> and <span class="math inline">\(R\)</span>, this suggests that there was some gene flow between <span class="math inline">\(T\)</span> and <span class="math inline">\(R\)</span>. Thus <span class="math inline">\(T\)</span> cannot be modeled as a combination of populations in <span class="math inline">\(L\)</span>. This may hint at the existence of a yet undiscovered “ghost” population which is related to a population in <span class="math inline">\(R\)</span>, but not represented by any of the populations in <span class="math inline">\(L\)</span>. On the other hand, if both models have the same rank (no extra gene flow between <span class="math inline">\(T\)</span> and <span class="math inline">\(R\)</span>), then <span class="math inline">\(T\)</span> can be modeled as a combination of populations in <span class="math inline">\(L\)</span>, and we can estimate the relative contribution of each population.</p>
<p><em>qpAdm</em> therefore produces two types of output that are of main interest: A p-value which tests whether there is a connection between <span class="math inline">\(R\)</span> and <span class="math inline">\(T\)</span>, other than through <span class="math inline">\(L\)</span>, and (if the answer to the last question is no), admixture weights which can tell us in what proportions populations in <span class="math inline">\(L\)</span> have contributed to <span class="math inline">\(T\)</span>.</p>
<p>The null hypothesis tested by <em>qpAdm</em> is that the populations in <span class="math inline">\(L\)</span> can account for all shared drift between <span class="math inline">\(T\)</span> and <span class="math inline">\(R\)</span>. A significant p-value rejecting this null hypothesis indicates that there has been gene flow between <span class="math inline">\(T\)</span> and <span class="math inline">\(R\)</span> (in either direction) that is not accounted for by any population in <span class="math inline">\(L\)</span>.</p>
<p>The admixture weights are estimated by computing the left null space of the matrix <span class="math inline">\(A\)</span> (they satisfy <span class="math inline">\(w A = 0\)</span>). This is basically asking in what proportions the populations in <span class="math inline">\(L\)</span> need to be added up so that their <span class="math inline">\(f_4\)</span>-statistics will be equal to the <span class="math inline">\(f_4\)</span>-statistics involving <span class="math inline">\(T\)</span>. These weights will also be computed when the model is rejected (the p-value is significant), but in that case they can’t be interpreted as admixture weights. It is important to keep in mind that the true source populations of <span class="math inline">\(T\)</span> are almost certainly not identical to any of the populations in <span class="math inline">\(L\)</span>, and that any inferences about the mixture proportions of the true, unobserved source populations depend on an unverifiable assumption: Each left population needs to be strictly cladal with one of the source populations.</p>
<!-- There are simpler methods which can model a population as a mixture of other populations. For example, variations of PCA or ADMIXTURE can model allele frequencies in one population as a linear combination of allele frequencies in other populations. These methods do not require a set of right populations, so why are they necessary in the qpAdm framework? The right populations serve two purposes: First, they allow us to test whether the L is sufficient to explain all variation in T, or whether there is evidence of shared drift between T and R which is not present in L. Second, they make it possible to assign the correct weights to populations in L even in cases where the populations in L have a complex demographic history. For this it is necessary that the populations in R are differentially related to the populations in L. -->
<!-- In most cases the sampled left populations L will not be the true source populations for T. However, the ancestry proportions of the true source populations can still be estimated accurately if each true source population forms a clade with its representative in L relative to all other modelled populations, and if the right populations are chosen such that no two left populations form a clade relative to all right populations. While the second assumption can be tested, the first assumption is untestable because it depends on potentially unobserved populations, which should be considered when interpreting qpAdm results. -->
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Robert Maier, Nick Patterson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'c30b6be96312414d83a0ba52b94b558b',
    indexName: 'admixtools',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
