<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Compute and store blocked allele frequency data — extract_afs_simple • admixtools</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous"><link href="../docsearch.css" rel="stylesheet"><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><script src="../extra.js"></script><meta property="og:title" content="Compute and store blocked allele frequency data — extract_afs_simple"><meta property="og:description" content="Prepare data for various ADMIXTOOLS 2 functions. Reads data from packedancestrymap or PLINK files,
and computes allele frequencies for selected populations and stores it as .rds files in outdir."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-9DRRL9PZD0"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-9DRRL9PZD0');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">admixtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/admixtools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/fstats.html">f-statistics</a>
    </li>
    <li>
      <a href="../articles/graphs.html">Admixture graphs</a>
    </li>
    <li>
      <a href="../articles/io.html">Data formats</a>
    </li>
    <li>
      <a href="../articles/paper.html">Paper notes</a>
    </li>
    <li>
      <a href="../articles/parallel.html">Parallelization</a>
    </li>
    <li>
      <a href="../articles/plotting.html">Plotting</a>
    </li>
    <li>
      <a href="../articles/qpadm.html">qpWave and qpAdm</a>
    </li>
    <li>
      <a href="../articles/resampling.html">Standard errors</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/uqrmaie1/admixtools/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul><form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off"></div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compute and store blocked allele frequency data</h1>
    <small class="dont-index">Source: <a href="https://github.com/uqrmaie1/admixtools/blob/HEAD/R/io.R" class="external-link"><code>R/io.R</code></a></small>
    <div class="hidden name"><code>extract_afs_simple.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Prepare data for various <em>ADMIXTOOLS 2</em> functions. Reads data from packedancestrymap or PLINK files,
and computes allele frequencies for selected populations and stores it as <code>.rds</code> files in outdir.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">extract_afs_simple</span><span class="op">(</span></span>
<span>  <span class="va">pref</span>,</span>
<span>  <span class="va">outdir</span>,</span>
<span>  inds <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pops <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  blgsize <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  cols_per_chunk <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  maxmiss <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  minmaf <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  maxmaf <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  minac2 <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  outpop <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transitions <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  transversions <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  keepsnps <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  format <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  poly_only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  adjust_pseudohaploid <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>pref</dt>
<dd><p>Prefix of <em>PLINK/EIGENSTRAT/PACKEDANCESTRYMAP</em> files.
<em>EIGENSTRAT/PACKEDANCESTRYMAP</em> have to end in <code>.geno</code>, <code>.snp</code>, <code>.ind</code>, <em>PLINK</em> has to end in <code>.bed</code>, <code>.bim</code>, <code>.fam</code></p></dd>


<dt>outdir</dt>
<dd><p>Directory where data will be stored.</p></dd>


<dt>inds</dt>
<dd><p>Individuals for which data should be extracted</p></dd>


<dt>pops</dt>
<dd><p>Populations for which data should be extracted. If both <code>pops</code> and <code>inds</code> are provided, they should have the same length and will be matched by position. If only <code>pops</code> is provided, all individuals from the <code>.ind</code> or <code>.fam</code> file in those populations will be extracted. If only <code>inds</code> is provided, each indivdual will be assigned to its own population of the same name. If neither <code>pops</code> nor <code>inds</code> is provided, all individuals and populations in the <code>.ind</code> or <code>.fam</code> file will be extracted.</p></dd>


<dt>blgsize</dt>
<dd><p>SNP block size in Morgan. Default is 0.05 (5 cM). If <code>blgsize</code> is 100 or greater, if will be interpreted as base pair distance rather than centimorgan distance.</p></dd>


<dt>cols_per_chunk</dt>
<dd><p>Number of populations per chunk. Lowering this number will lower the memory requirements when running <code><a href="afs_to_f2.html">afs_to_f2</a></code>, but more chunk pairs will have to be computed.</p></dd>


<dt>maxmiss</dt>
<dd><p>Discard SNPs which are missing in a fraction of populations higher than <code>maxmiss</code></p></dd>


<dt>minmaf</dt>
<dd><p>Discard SNPs with minor allele frequency less than <code>minmaf</code></p></dd>


<dt>maxmaf</dt>
<dd><p>Discard SNPs with minor allele frequency greater than than <code>maxmaf</code></p></dd>


<dt>minac2</dt>
<dd><p>Discard SNPs with allele count lower than 2 in any population (default <code>FALSE</code>). This option should be set to <code>TRUE</code> when computing f3-statistics where one population consists mostly of pseudohaploid samples. Otherwise heterozygosity estimates and thus f3-estimates can be biased. <code>minac2 == 2</code> will discard SNPs with allele count lower than 2 in any non-singleton population (this option is experimental and is based on the hypothesis that using SNPs with allele count lower than 2 only leads to biases in non-singleton populations). While the <code>minac2</code> option discards SNPs with allele count lower than 2 in any population, the <code><a href="qp3pop.html">qp3pop</a></code> function will only discard SNPs with allele count lower than 2 in the first (target) population (when the first argument is the prefix of a genotype file).</p></dd>


<dt>outpop</dt>
<dd><p>Keep only SNPs which are heterozygous in this population</p></dd>


<dt>transitions</dt>
<dd><p>Set this to <code>FALSE</code> to exclude transition SNPs</p></dd>


<dt>transversions</dt>
<dd><p>Set this to <code>FALSE</code> to exclude transversion SNPs</p></dd>


<dt>keepsnps</dt>
<dd><p>SNP IDs of SNPs to keep. Overrides other SNP filtering options</p></dd>


<dt>format</dt>
<dd><p>Supply this if the prefix can refer to genotype data in different formats
and you want to choose which one to read. Should be <code>plink</code> to read <code>.bed</code>, <code>.bim</code>, <code>.fam</code> files, or <code>eigenstrat</code>, or <code>packedancestrymap</code> to read <code>.geno</code>, <code>.snp</code>, <code>.ind</code> files.</p></dd>


<dt>poly_only</dt>
<dd><p>Specify whether SNPs with identical allele frequencies in every population should be discarded (<code>poly_only = TRUE</code>), or whether they should be used (<code>poly_only = FALSE</code>). By default (<code>poly_only = c("f2")</code>), these SNPs will be used to compute FST and allele frequency products, but not to compute f2 (this is the default option in the original ADMIXTOOLS).</p></dd>


<dt>adjust_pseudohaploid</dt>
<dd><p>Genotypes of pseudohaploid samples are usually coded as <code>0</code> or <code>2</code>, even though only one allele is observed. <code>adjust_pseudohaploid</code> ensures that the observed allele count increases only by <code>1</code> for each pseudohaploid sample. If <code>TRUE</code> (default), samples that don't have any genotypes coded as <code>1</code> among the first 1000 SNPs are automatically identified as pseudohaploid. This leads to slightly more accurate estimates of f-statistics. Setting this parameter to <code>FALSE</code> treats all samples as diploid and is equivalent to the <em>ADMIXTOOLS</em> <code>inbreed: NO</code> option. Setting <code>adjust_pseudohaploid</code> to an integer <code>n</code> will check the first <code>n</code> SNPs instead of the first 1000 SNPs.</p></dd>


<dt>verbose</dt>
<dd><p>Print progress updates</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>SNP metadata (invisibly)</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="va">pref</span> <span class="op">=</span> <span class="st">'my/genofiles/prefix'</span></span></span>
<span class="r-in"><span><span class="va">outdir</span> <span class="op">=</span> <span class="st">'dir/for/afdata/'</span></span></span>
<span class="r-in"><span><span class="fu"><a href="extract_afs.html">extract_afs</a></span><span class="op">(</span><span class="va">pref</span>, <span class="va">outdir</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Robert Maier, Nick Patterson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'c30b6be96312414d83a0ba52b94b558b',
    indexName: 'admixtools',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script></body></html>

