<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ admixtools</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>



  <script src="../extra.js"></script>

<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-9DRRL9PZD0"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-9DRRL9PZD0');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">admixtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/admixtools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/fstats.html">f-statistics</a>
    </li>
    <li>
      <a href="../articles/graphs.html">Admixture graphs</a>
    </li>
    <li>
      <a href="../articles/io.html">Data formats</a>
    </li>
    <li>
      <a href="../articles/paper.html">Paper notes</a>
    </li>
    <li>
      <a href="../articles/parallel.html">Parallelization</a>
    </li>
    <li>
      <a href="../articles/plotting.html">Plotting</a>
    </li>
    <li>
      <a href="../articles/qpadm.html">qpWave and qpAdm</a>
    </li>
    <li>
      <a href="../articles/resampling.html">Standard errors</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/uqrmaie1/admixtools/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-admixtools-programs" class="hasAnchor"><a href="#section-admixtools-programs" class="anchor"></a>ADMIXTOOLS programs</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="qp3pop.html">qp3pop()</a></code> </p>
        </td>
        <td><p>Estimate f3 statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpdstat.html">qpdstat()</a></code> </p>
        </td>
        <td><p>Estimate f4 statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpf4ratio.html">qpf4ratio()</a></code> </p>
        </td>
        <td><p>Estimate admixture proportions via f4 ratios</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpwave.html">qpwave()</a></code> </p>
        </td>
        <td><p>Estimate admixture waves</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpadm.html">qpadm()</a></code> </p>
        </td>
        <td><p>Estimate admixture weights</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpgraph.html">qpgraph()</a></code> </p>
        </td>
        <td><p>Compute the fit of an admixture graph</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-wrapper-functions" class="hasAnchor"><a href="#section-wrapper-functions" class="anchor"></a>Wrapper functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="qp3pop_wrapper.html">qp3pop_wrapper()</a></code> </p>
        </td>
        <td><p>Wrapper function around the original qp3Pop program</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpadm_wrapper.html">qpadm_wrapper()</a></code> </p>
        </td>
        <td><p>Wrapper function around the original qpAdm program</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpdstat_wrapper.html">qpdstat_wrapper()</a></code> </p>
        </td>
        <td><p>Wrapper function around the original qpDstat program</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpf4ratio_wrapper.html">qpf4ratio_wrapper()</a></code> </p>
        </td>
        <td><p>Wrapper function around the original qpF4ratio program</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpgraph_wrapper.html">qpgraph_wrapper()</a></code> </p>
        </td>
        <td><p>Wrapper function around the original qpGraph program</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpwave_wrapper.html">qpwave_wrapper()</a></code> </p>
        </td>
        <td><p>Wrapper function around the original qpWave program</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-parsing-functions" class="hasAnchor"><a href="#section-parsing-functions" class="anchor"></a>Parsing functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="parse_dot.html">parse_dot()</a></code> </p>
        </td>
        <td><p>Read graph in dot format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qp3pop_output.html">parse_qp3pop_output()</a></code> </p>
        </td>
        <td><p>Read qp3Pop output file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qpadm_output.html">parse_qpadm_output()</a></code> </p>
        </td>
        <td><p>Read qpAdm output file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qpdstat_output.html">parse_qpdstat_output()</a></code> </p>
        </td>
        <td><p>Read qpDstat output file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qpf4ratio_output.html">parse_qpf4ratio_output()</a></code> </p>
        </td>
        <td><p>Read qpF4ratio output file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qpgraph_graphfile.html">parse_qpgraph_graphfile()</a></code> </p>
        </td>
        <td><p>Read qpGraph graph file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qpgraph_output.html">parse_qpgraph_output()</a></code> </p>
        </td>
        <td><p>Read qpGraph output file</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-plotting-functions" class="hasAnchor"><a href="#section-plotting-functions" class="anchor"></a>Plotting functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="plot_comparison.html">plot_comparison()</a></code> </p>
        </td>
        <td><p>Compare two models</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_graph.html">plot_graph()</a></code> </p>
        </td>
        <td><p>Plot an admixture graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_graph_map.html">plot_graph_map()</a></code> </p>
        </td>
        <td><p>Plot an admixture graph on a map</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_map.html">plot_map()</a></code> </p>
        </td>
        <td><p>Plot samples on a map</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotly_comparison.html">plotly_comparison()</a></code> </p>
        </td>
        <td><p>Compare two models</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotly_graph.html">plotly_graph()</a></code> </p>
        </td>
        <td><p>Plot an admixture graph using plotly</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-resampling-functions" class="hasAnchor"><a href="#section-resampling-functions" class="anchor"></a>Resampling functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="resample_inds.html">qp3pop_resample_inds()</a></code> <code><a href="resample_inds.html">qpdstat_resample_inds()</a></code> <code><a href="resample_inds.html">qpwave_resample_inds()</a></code> <code><a href="resample_inds.html">qpadm_resample_inds()</a></code> <code><a href="resample_inds.html">qpgraph_resample_inds()</a></code> </p>
        </td>
        <td><p>Run models while leaving out individuals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="resample_snps.html">qp3pop_resample_snps()</a></code> <code><a href="resample_snps.html">qpdstat_resample_snps()</a></code> <code><a href="resample_snps.html">qpwave_resample_snps()</a></code> <code><a href="resample_snps.html">qpadm_resample_snps()</a></code> <code><a href="resample_snps.html">qpgraph_resample_snps()</a></code> </p>
        </td>
        <td><p>Run models while leaving out SNP blocks</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-reading-and-writing" class="hasAnchor"><a href="#section-reading-and-writing" class="anchor"></a>Reading and writing</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="extract_f2.html">extract_f2()</a></code> </p>
        </td>
        <td><p>Compute and store blocked f2 statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extract_counts.html">extract_counts()</a></code> </p>
        </td>
        <td><p>Extract and store data needed to compute blocked f2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="f2_from_precomp.html">f2_from_precomp()</a></code> </p>
        </td>
        <td><p>Read blocked f2 statistics from disk</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_packedancestrymap.html">read_packedancestrymap()</a></code> </p>
        </td>
        <td><p>Read genotype data from packedancestrymap files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_plink.html">read_plink()</a></code> </p>
        </td>
        <td><p>Read genotype data from <code>PLINK</code> files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="packedancestrymap_to_afs.html">packedancestrymap_to_afs()</a></code> </p>
        </td>
        <td><p>Read allele frequencies from packedancestrymap files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plink_to_afs.html">plink_to_afs()</a></code> </p>
        </td>
        <td><p>Read allele frequencies from <code>PLINK</code> files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="packedancestrymap_to_plink.html">packedancestrymap_to_plink()</a></code> </p>
        </td>
        <td><p>Convert <em>EIGENSTRAT</em> or <em>PACKEDANCESTRYMAP</em> to <em>PLINK</em></p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-model-exploration" class="hasAnchor"><a href="#section-model-exploration" class="anchor"></a>Model exploration</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="find_graphs.html">find_graphs()</a></code> </p>
        </td>
        <td><p>Find well fitting admixture graphs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpwave_pairs.html">qpwave_pairs()</a></code> </p>
        </td>
        <td><p>Compute all pairwise qpwave p-values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpadm_rotate.html">qpadm_rotate()</a></code> </p>
        </td>
        <td><p>Compute p-values for many qpadm models</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-everything" class="hasAnchor"><a href="#section-everything" class="anchor"></a>Everything</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="admixtools.html">admixtools</a></code> </p>
        </td>
        <td><p>Tools for inferring demographic history from genetic data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="afs_to_counts.html">afs_to_counts()</a></code> </p>
        </td>
        <td><p>Compute count blocks and write them to disk</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="afs_to_f2.html">afs_to_f2()</a></code> </p>
        </td>
        <td><p>Compute f2 blocks and write them to disk</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="afs_to_f2_blocks.html">afs_to_f2_blocks()</a></code> </p>
        </td>
        <td><p>Compute all pairwise f2 statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="agraph_to_igraph.html">agraph_to_igraph()</a></code> </p>
        </td>
        <td><p>Convert agraph to igraph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="boo_list.html">boo_list()</a></code> </p>
        </td>
        <td><p>Generate a list of bootstrap resampled arrays</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compare_fits.html">compare_fits()</a></code> </p>
        </td>
        <td><p>Compare the fit of two qpgraph models</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compare_fits2.html">compare_fits2()</a></code> </p>
        </td>
        <td><p>Compare the fit of two qpgraph models</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compare_fits4.html">compare_fits4()</a></code> </p>
        </td>
        <td><p>Compare the fit of two qpgraph models</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="count_snps.html">count_snps()</a></code> </p>
        </td>
        <td><p>Count SNPs in an f2-statistics array</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="count_zero_edges.html">count_zero_edges()</a></code> </p>
        </td>
        <td><p>Count zero-length edges</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="decomposed_tree_neighbors.html">decomposed_tree_neighbors()</a></code> </p>
        </td>
        <td><p>Find all trees within SPR distance of 1 of all graph component trees</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="delete_admix.html">delete_admix()</a></code> </p>
        </td>
        <td><p>Delete an admixture edge</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="delete_groups.html">delete_groups()</a></code> </p>
        </td>
        <td><p>Delete groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="delete_leaf.html">delete_leaf()</a></code> </p>
        </td>
        <td><p>Remove population from graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="desimplify_graph.html">desimplify_graph()</a></code> </p>
        </td>
        <td><p>Add two nodes before each admixture node</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="edges_to_igraph.html">edges_to_igraph()</a></code> </p>
        </td>
        <td><p>Convert data frame graph to igraph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="eigenstrat_to_afs.html">eigenstrat_to_afs()</a></code> </p>
        </td>
        <td><p>Read allele frequencies from <em>EIGENSTRAT</em> files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="est_to_boo.html">est_to_boo()</a></code> </p>
        </td>
        <td><p>Turn per-block estimates into bootstrap estimates</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="est_to_loo.html">est_to_loo()</a></code> </p>
        </td>
        <td><p>Turn per-block estimates into leave-one-out estimates</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="example_anno.html">example_anno</a></code> </p>
        </td>
        <td><p>Data frame with sample annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="example_f2_blocks.html">example_f2_blocks</a></code> </p>
        </td>
        <td><p>Blocked f2-statistics for 7 populations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="example_graph.html">example_graph</a></code> </p>
        </td>
        <td><p>Admixture graph for 7 populations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="example_igraph.html">example_igraph</a></code> </p>
        </td>
        <td><p>Admixture graph for 7 populations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="example_opt.html">example_opt</a></code> </p>
        </td>
        <td><p>Data frame with one fitted admixture graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="example_qpgraph_ref_results.html">example_qpgraph_ref_results</a></code> </p>
        </td>
        <td><p>example_graph fitted using qpGraph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="example_triples.html">example_triples</a></code> </p>
        </td>
        <td><p>Data frame with population triples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extract_afs.html">extract_afs()</a></code> </p>
        </td>
        <td><p>Compute and store blocked allele frequency data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extract_afs_simple.html">extract_afs_simple()</a></code> </p>
        </td>
        <td><p>Compute and store blocked allele frequency data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extract_counts.html">extract_counts()</a></code> </p>
        </td>
        <td><p>Extract and store data needed to compute blocked f2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extract_f2.html">extract_f2()</a></code> </p>
        </td>
        <td><p>Compute and store blocked f2 statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extract_f2_large.html">extract_f2_large()</a></code> </p>
        </td>
        <td><p>Compute and store blocked f2 statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extract_f2_subset.html">extract_f2_subset()</a></code> </p>
        </td>
        <td><p>Copy f2-statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extract_samples.html">extract_samples()</a></code> </p>
        </td>
        <td><p>Extract samples from PLINK files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="f2.html">f2()</a></code> </p>
        </td>
        <td><p>Estimate f2 statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="f2_from_geno.html">f2_from_geno()</a></code> </p>
        </td>
        <td><p>Compute blocked f2 statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="f2_from_msprime.html">f2_from_msprime()</a></code> </p>
        </td>
        <td><p>Simulate an admixture graph in msprime</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="f2_from_precomp.html">f2_from_precomp()</a></code> </p>
        </td>
        <td><p>Read blocked f2 statistics from disk</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="f2dat_f4dat.html">f2dat_f4dat()</a></code> </p>
        </td>
        <td><p>Turn f2 data to f4 data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="f3blockdat_from_geno.html">f3blockdat_from_geno()</a></code> </p>
        </td>
        <td><p>f3 from genotype data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="f4_from_afdat.html">f4_from_afdat()</a></code> </p>
        </td>
        <td><p>Compute f4 from allele frequencies</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="f4_from_f2.html">f4_from_f2()</a></code> </p>
        </td>
        <td><p>Get per-block f4-statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="f4blockdat_from_geno.html">f4blockdat_from_geno()</a></code> </p>
        </td>
        <td><p>f4 from genotype data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="f4blockdat_to_f4blocks.html">f4blockdat_to_f4blocks()</a></code> </p>
        </td>
        <td><p>Turn f4 block data to 3d array</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="find_admixedges.html">find_admixedges()</a></code> </p>
        </td>
        <td><p>Find admixture edges</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="find_graphs.html">find_graphs()</a></code> </p>
        </td>
        <td><p>Find well fitting admixture graphs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="find_graphs_old.html">find_graphs_old()</a></code> </p>
        </td>
        <td><p>Find well fitting admixture graphs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="find_newedges.html">find_newedges()</a></code> </p>
        </td>
        <td><p>Find possible new edges</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="find_normedges.html">find_normedges()</a></code> </p>
        </td>
        <td><p>Find drift edges</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="flipadmix_random.html">flipadmix_random()</a></code> </p>
        </td>
        <td><p>Modify a graph flipping the direction of an admixture edge</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fst.html">fst()</a></code> </p>
        </td>
        <td><p>Compute Fst</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="generate_all_graphs.html">generate_all_graphs()</a></code> </p>
        </td>
        <td><p>Generate all graphs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="generate_all_trees.html">generate_all_trees()</a></code> </p>
        </td>
        <td><p>Generate all trees</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_block_lengths.html">get_block_lengths()</a></code> </p>
        </td>
        <td><p>Find LD-independent blocks</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_f2.html">get_f2()</a></code> </p>
        </td>
        <td><p>Turns f2_data into f2_blocks</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_leafnames.html">get_leafnames()</a></code> </p>
        </td>
        <td><p>Get the population names of a graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_outpop.html">get_outpop()</a></code> </p>
        </td>
        <td><p>Get the outgroup from a graph (if it exists)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_rootname.html">get_rootname()</a></code> </p>
        </td>
        <td><p>Get the root name</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="graph_addleaf.html">graph_addleaf()</a></code> </p>
        </td>
        <td><p>Add a population to an admixture graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="graph_distances.html">graph_distances()</a></code> </p>
        </td>
        <td><p>Pairwise distance estimates for graphs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="graph_equations.html">graph_equations()</a></code> </p>
        </td>
        <td><p>Find well fitting admixture graphs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="graph_f2_function.html">graph_f2_function()</a></code> </p>
        </td>
        <td><p>Make a function representing a graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="graph_flipadmix.html">graph_flipadmix()</a></code> </p>
        </td>
        <td><p>Find all valid graphs which result from flipping one admixture edge</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="graph_hash.html">graph_hash()</a></code> </p>
        </td>
        <td><p>Get unique hash of an admixture graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="graph_minusone.html">graph_minusone()</a></code> </p>
        </td>
        <td><p>Find all graphs which result from removing one admixture edge</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="graph_minusplus.html">graph_minusplus()</a></code> </p>
        </td>
        <td><p>Find all graphs which result from adding and removing one admixture edge</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="graph_plusone.html">graph_plusone()</a></code> </p>
        </td>
        <td><p>Find all graphs which result from adding one admixture edge</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="graph_splittrees.html">graph_splittrees()</a></code> </p>
        </td>
        <td><p>Find all trees which are part of the admixture graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="graph_to_lgo.html">graph_to_lgo()</a></code> </p>
        </td>
        <td><p>Convert a graph to .lgo file to be used in Legofit software</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="graph_to_qpadm.html">graph_to_qpadm()</a></code> </p>
        </td>
        <td><p>Get all qpadm models for a graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="graphmod_pavel.html">graphmod_pavel()</a></code> </p>
        </td>
        <td><p>Return all graphs created from permuting a subclade</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="group_samples.html">group_samples()</a></code> </p>
        </td>
        <td><p>Group precomputed data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="igraph_to_agraph.html">igraph_to_agraph()</a></code> </p>
        </td>
        <td><p>Convert igraph to agraph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="insert_admix.html">insert_admix()</a></code> </p>
        </td>
        <td><p>Insert a single edge into graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="insert_admix_n.html">insert_admix_n()</a></code> </p>
        </td>
        <td><p>Insert admixture edges into graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="insert_admix_old.html">insert_admix_old()</a></code> </p>
        </td>
        <td><p>Insert admixture edges into graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="insert_leaf.html">insert_leaf()</a></code> </p>
        </td>
        <td><p>Add population to graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="is_valid.html">is_valid()</a></code> </p>
        </td>
        <td><p>Test if an admixture graph is valid</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="isomorphism_classes.html">isomorphism_classes()</a></code> </p>
        </td>
        <td><p>Find identical graphs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="isomorphism_classes2.html">isomorphism_classes2()</a></code> </p>
        </td>
        <td><p>Find identical graphs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="joint_sfs.html">joint_sfs()</a></code> </p>
        </td>
        <td><p>Joint site frequency spectrum</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="joint_spectrum.html">joint_spectrum()</a></code> </p>
        </td>
        <td><p>Estimate joint allele frequency spectrum</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lazadm.html">lazadm()</a></code> </p>
        </td>
        <td><p>Estimate admixture weights</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loo_list.html">loo_list()</a></code> </p>
        </td>
        <td><p>Generate a list of leave-one-out arrays</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loo_to_est.html">loo_to_est()</a></code> </p>
        </td>
        <td><p>Turn leave-one-out estimates to per-block estimates</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="move_admixedge_once.html">move_admixedge_once()</a></code> </p>
        </td>
        <td><p>Modify a graph by moving an admixture edge</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="msprime_genome.html">msprime_genome()</a></code> </p>
        </td>
        <td><p>Simulate an admixture graph in msprime v1.x.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="msprime_sim.html">msprime_sim()</a></code> </p>
        </td>
        <td><p>Simulate an admixture graph in msprime v1.x</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mutate_n.html">mutate_n()</a></code> </p>
        </td>
        <td><p>Modify a graph by applying n mutation functions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="newick_to_edges.html">newick_to_edges()</a></code> </p>
        </td>
        <td><p>Turn a newick format tree to a matrix of edges</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="node_counts.html">node_counts()</a></code> </p>
        </td>
        <td><p>Count how often each node in graph occurs in other graphs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="node_signature.html">node_signature()</a></code> </p>
        </td>
        <td><p>Returns a signature of a graph consisting of the left and right descendent leaf nodes of each internal node (sorted and concatenated)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="numadmix.html">numadmix()</a></code> </p>
        </td>
        <td><p>Count number of admixture nodes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="packedancestrymap_to_afs.html">packedancestrymap_to_afs()</a></code> </p>
        </td>
        <td><p>Read allele frequencies from packedancestrymap files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="packedancestrymap_to_plink.html">packedancestrymap_to_plink()</a></code> </p>
        </td>
        <td><p>Convert <em>EIGENSTRAT</em> or <em>PACKEDANCESTRYMAP</em> to <em>PLINK</em></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_dot.html">parse_dot()</a></code> </p>
        </td>
        <td><p>Read graph in dot format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qp3pop_output.html">parse_qp3pop_output()</a></code> </p>
        </td>
        <td><p>Read qp3Pop output file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qpadm_output.html">parse_qpadm_output()</a></code> </p>
        </td>
        <td><p>Read qpAdm output file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qpdstat_output.html">parse_qpdstat_output()</a></code> </p>
        </td>
        <td><p>Read qpDstat output file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qpf4ratio_output.html">parse_qpf4ratio_output()</a></code> </p>
        </td>
        <td><p>Read qpF4ratio output file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qpgraph_graphfile.html">parse_qpgraph_graphfile()</a></code> </p>
        </td>
        <td><p>Read qpGraph graph file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qpgraph_output.html">parse_qpgraph_output()</a></code> </p>
        </td>
        <td><p>Read qpGraph output file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="permute_leaves.html">permute_leaves()</a></code> </p>
        </td>
        <td><p>Modify a graph by permuting leaf nodes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="place_root_random.html">place_root_random()</a></code> </p>
        </td>
        <td><p>Modify a graph by changing the position of the root node</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plink_to_afs.html">plink_to_afs()</a></code> </p>
        </td>
        <td><p>Read allele frequencies from <code>PLINK</code> files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_comparison.html">plot_comparison()</a></code> </p>
        </td>
        <td><p>Compare two models</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_graph.html">plot_graph()</a></code> </p>
        </td>
        <td><p>Plot an admixture graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_graph_map.html">plot_graph_map()</a></code> </p>
        </td>
        <td><p>Plot an admixture graph on a map</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_map.html">plot_map()</a></code> </p>
        </td>
        <td><p>Plot samples on a map</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotly_comparison.html">plotly_comparison()</a></code> </p>
        </td>
        <td><p>Compare two models</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotly_graph.html">plotly_graph()</a></code> </p>
        </td>
        <td><p>Plot an admixture graph using plotly</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pseudo_dates.html">pseudo_dates()</a></code> </p>
        </td>
        <td><p>Get pseudo dates for graph nodes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qp3pop.html">qp3pop()</a></code> </p>
        </td>
        <td><p>Estimate f3 statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qp3pop_wrapper.html">qp3pop_wrapper()</a></code> </p>
        </td>
        <td><p>Wrapper function around the original qp3Pop program</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpadm.html">qpadm()</a></code> </p>
        </td>
        <td><p>Estimate admixture weights</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpadm_models.html">qpadm_models()</a></code> </p>
        </td>
        <td><p>Partition a list of populations into left and right populations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpadm_models_old.html">qpadm_models_old()</a></code> </p>
        </td>
        <td><p>Return all valid qpAdm models for an admixturegraph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpadm_multi.html">qpadm_multi()</a></code> </p>
        </td>
        <td><p>Run multiple qpadm models</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpadm_p.html">qpadm_p()</a></code> </p>
        </td>
        <td><p>Faster version of <code><a href='https://rdrr.io/pkg/admixtools/man/qpadm.html'>qpadm</a></code> with reduced output</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpadm_rotate.html">qpadm_rotate()</a></code> </p>
        </td>
        <td><p>Compute p-values for many qpadm models</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpadm_wrapper.html">qpadm_wrapper()</a></code> </p>
        </td>
        <td><p>Wrapper function around the original qpAdm program</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpdstat.html">qpdstat()</a></code> </p>
        </td>
        <td><p>Estimate f4 statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpdstat_wrapper.html">qpdstat_wrapper()</a></code> </p>
        </td>
        <td><p>Wrapper function around the original qpDstat program</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpf4ratio.html">qpf4ratio()</a></code> </p>
        </td>
        <td><p>Estimate admixture proportions via f4 ratios</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpf4ratio_wrapper.html">qpf4ratio_wrapper()</a></code> </p>
        </td>
        <td><p>Wrapper function around the original qpF4ratio program</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpfstats.html">qpfstats()</a></code> </p>
        </td>
        <td><p>Get smoothed f2-statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpgraph.html">qpgraph()</a></code> </p>
        </td>
        <td><p>Compute the fit of an admixture graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpgraph_precompute_f3.html">qpgraph_precompute_f3()</a></code> </p>
        </td>
        <td><p>Compute f3-statistics from f2-statistics.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpgraph_resample_multi.html">qpgraph_resample_multi()</a></code> </p>
        </td>
        <td><p>Evaluate a qpgraph models many times</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpgraph_resample_snps2.html">qpgraph_resample_snps2()</a></code> </p>
        </td>
        <td><p>Evaluate a qpgraph model many times</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpgraph_wrapper.html">qpgraph_wrapper()</a></code> </p>
        </td>
        <td><p>Wrapper function around the original qpGraph program</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpwave.html">qpwave()</a></code> </p>
        </td>
        <td><p>Estimate admixture waves</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpwave_pairs.html">qpwave_pairs()</a></code> </p>
        </td>
        <td><p>Compute all pairwise qpwave p-values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpwave_wrapper.html">qpwave_wrapper()</a></code> </p>
        </td>
        <td><p>Wrapper function around the original qpWave program</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="random_admixturegraph.html">random_admixturegraph()</a></code> </p>
        </td>
        <td><p>Generate a random admixture graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="random_dates.html">random_dates()</a></code> </p>
        </td>
        <td><p>Get random dates for graph nodes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="random_newick.html">random_newick()</a></code> </p>
        </td>
        <td><p>Generate a random binary graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="random_sim.html">random_sim()</a></code> </p>
        </td>
        <td><p>Generate a random graph and simulate it in msprime v1.x</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_eigenstrat.html">read_eigenstrat()</a></code> </p>
        </td>
        <td><p>Read genotype data from <em>EIGENSTRAT</em> files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_f2.html">read_f2()</a></code> </p>
        </td>
        <td><p>Read blocked f2 estimates from disk</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_packedancestrymap.html">read_packedancestrymap()</a></code> </p>
        </td>
        <td><p>Read genotype data from packedancestrymap files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_plink.html">read_plink()</a></code> </p>
        </td>
        <td><p>Read genotype data from <code>PLINK</code> files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="resample_inds.html">qp3pop_resample_inds()</a></code> <code><a href="resample_inds.html">qpdstat_resample_inds()</a></code> <code><a href="resample_inds.html">qpwave_resample_inds()</a></code> <code><a href="resample_inds.html">qpadm_resample_inds()</a></code> <code><a href="resample_inds.html">qpgraph_resample_inds()</a></code> </p>
        </td>
        <td><p>Run models while leaving out individuals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="resample_snps.html">qp3pop_resample_snps()</a></code> <code><a href="resample_snps.html">qpdstat_resample_snps()</a></code> <code><a href="resample_snps.html">qpwave_resample_snps()</a></code> <code><a href="resample_snps.html">qpadm_resample_snps()</a></code> <code><a href="resample_snps.html">qpgraph_resample_snps()</a></code> </p>
        </td>
        <td><p>Run models while leaving out SNP blocks</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rotate_models.html">rotate_models()</a></code> </p>
        </td>
        <td><p>Rotate populations between left and right</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_shiny_admixtools.html">run_shiny_admixtools()</a></code> </p>
        </td>
        <td><p>Launch ADMIXTOOLS 2 GUI</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="satisfies_constraints.html">satisfies_constraints()</a></code> </p>
        </td>
        <td><p>Test constraints on a graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="satisfies_eventorder.html">satisfies_eventorder()</a></code> </p>
        </td>
        <td><p>Test f4 constraints on a graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="satisfies_nonzerof4.html">satisfies_nonzerof4()</a></code> </p>
        </td>
        <td><p>Test f4 constraints on a graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="satisfies_numadmix.html">satisfies_numadmix()</a></code> </p>
        </td>
        <td><p>Test admixture constraints on a graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="satisfies_zerof4.html">satisfies_zerof4()</a></code> </p>
        </td>
        <td><p>Test f4 constraints on a graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="shortest_unique_prefixes.html">shortest_unique_prefixes()</a></code> </p>
        </td>
        <td><p>Return shortest unique prefixes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="simplify_graph.html">simplify_graph()</a></code> </p>
        </td>
        <td><p>Remove redundant edges</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="split_mat.html">split_mat()</a></code> </p>
        </td>
        <td><p>Split a matrix into blocks</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="split_multifurcations.html">split_multifurcations()</a></code> </p>
        </td>
        <td><p>Split nodes with more than two edges</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spr_all.html">spr_all()</a></code> </p>
        </td>
        <td><p>Modify a graph by regrafting a subcomponent</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spr_leaves.html">spr_leaves()</a></code> </p>
        </td>
        <td><p>Modify a graph by regrafting a leaf</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarize_descendants.html">summarize_descendants()</a></code> </p>
        </td>
        <td><p>List leaf nodes for all internal nodes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarize_descendants_list.html">summarize_descendants_list()</a></code> </p>
        </td>
        <td><p>List leaf nodes for all internal nodes in a list of graphs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarize_eventorder.html">summarize_eventorder()</a></code> </p>
        </td>
        <td><p>List population split events in a graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarize_eventorder_list.html">summarize_eventorder_list()</a></code> </p>
        </td>
        <td><p>List population split events in a list of graphs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarize_fits.html">summarize_fits()</a></code> </p>
        </td>
        <td><p>Summarize graph fits</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarize_numadmix.html">summarize_numadmix()</a></code> </p>
        </td>
        <td><p>List number of admixture events for each population</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarize_numadmix_list.html">summarize_numadmix_list()</a></code> </p>
        </td>
        <td><p>List number of admixture events for each population in a list of graphs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarize_proxies.html">summarize_proxies()</a></code> </p>
        </td>
        <td><p>Assign proxy populations to admixed populations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarize_proxies_list.html">summarize_proxies_list()</a></code> </p>
        </td>
        <td><p>List proxy populations in graphlist</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarize_triples.html">summarize_triples()</a></code> </p>
        </td>
        <td><p>Summarize triples across graphs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarize_zerof4.html">summarize_zerof4()</a></code> </p>
        </td>
        <td><p>List clades in a graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarize_zerof4_list.html">summarize_zerof4_list()</a></code> </p>
        </td>
        <td><p>List clades in a list of graphs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="swap_leaves.html">swap_leaves()</a></code> </p>
        </td>
        <td><p>Modify a graph by swapping two leaf nodes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="test_cladality.html">test_cladality()</a></code> </p>
        </td>
        <td><p>Test if two sets of populations form two clades</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tree_in_graph.html">tree_in_graph()</a></code> </p>
        </td>
        <td><p>Test if a tree is part of a graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tree_neighbors.html">tree_neighbors()</a></code> </p>
        </td>
        <td><p>Find all trees within SPR distance of 1</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="unidentifiable_edges.html">unidentifiable_edges()</a></code> </p>
        </td>
        <td><p>Find all unidentifiable edges</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="write_dot.html">write_dot()</a></code> </p>
        </td>
        <td><p>Convert graph to dot format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="write_f2.html">write_f2()</a></code> </p>
        </td>
        <td><p>Write blocked f2 estimates to disk</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Robert Maier, Nick Patterson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: 'c30b6be96312414d83a0ba52b94b558b',
    indexName: 'admixtools',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  </body>
</html>


